<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gctree.dag_filtering &mdash; gctree  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="gctree.dag_filtering.filter_dag" href="gctree.dag_filtering.filter_dag.html" />
    <link rel="prev" title="gctree.isotyping.IsotypeTemplate" href="gctree.isotyping.IsotypeTemplate.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cite.html">Citing gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rendering-demo.html">Custom tree visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">CLI Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../cli.html">gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cli.html#parsimony-utilities">Parsimony utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cli.html#isotype">isotype</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../api.html">Classes</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html#modules">Modules</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="gctree.mutation_model.html">gctree.mutation_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="gctree.utils.html">gctree.utils</a></li>
<li class="toctree-l2"><a class="reference internal" href="gctree.isotyping.html">gctree.isotyping</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">gctree.dag_filtering</a><ul>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.filter_dag.html">gctree.dag_filtering.filter_dag</a></li>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.isotype_dagfuncs.html">gctree.dag_filtering.isotype_dagfuncs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.ll_cmcount_dagfuncs.html">gctree.dag_filtering.ll_cmcount_dagfuncs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.ll_genotype_dagfuncs.html">gctree.dag_filtering.ll_genotype_dagfuncs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.mutability_dagfuncs.html">gctree.dag_filtering.mutability_dagfuncs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gctree.dag_filtering.mutability_distance.html">gctree.dag_filtering.mutability_distance</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#developer-tools">Developer tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#todo-list">Todo list</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">gctree</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../api.html">Classes</a> &raquo;</li>
      <li>gctree.dag_filtering</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="module-gctree.dag_filtering">
<span id="gctree-dag-filtering"></span><h1>gctree.dag_filtering<a class="headerlink" href="#module-gctree.dag_filtering" title="Permalink to this headline"></a></h1>
<p>Functions for choosing optimal trees according to flexible criteria.</p>
<p class="rubric">Functions</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="gctree.dag_filtering.filter_dag.html#gctree.dag_filtering.filter_dag" title="gctree.dag_filtering.filter_dag"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_dag</span></code></a></p></td>
<td><p>Filter trees in the provided history DAG.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="gctree.dag_filtering.isotype_dagfuncs.html#gctree.dag_filtering.isotype_dagfuncs" title="gctree.dag_filtering.isotype_dagfuncs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">isotype_dagfuncs</span></code></a></p></td>
<td><p>Return functions for filtering by isotype parsimony score on the history DAG.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="gctree.dag_filtering.ll_cmcount_dagfuncs.html#gctree.dag_filtering.ll_cmcount_dagfuncs" title="gctree.dag_filtering.ll_cmcount_dagfuncs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ll_cmcount_dagfuncs</span></code></a></p></td>
<td><p>A slower but more numerically stable equivalent to ll_genotype_dagfuncs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="gctree.dag_filtering.ll_genotype_dagfuncs.html#gctree.dag_filtering.ll_genotype_dagfuncs" title="gctree.dag_filtering.ll_genotype_dagfuncs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ll_genotype_dagfuncs</span></code></a></p></td>
<td><p>Return functions for counting tree log likelihood on the history DAG.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="gctree.dag_filtering.mutability_dagfuncs.html#gctree.dag_filtering.mutability_dagfuncs" title="gctree.dag_filtering.mutability_dagfuncs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutability_dagfuncs</span></code></a></p></td>
<td><p>Return functions for counting mutability parsimony on the history DAG.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="gctree.dag_filtering.mutability_distance.html#gctree.dag_filtering.mutability_distance" title="gctree.dag_filtering.mutability_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutability_distance</span></code></a></p></td>
<td><p>Returns a fast distance function based on passed mutation_model.</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.filter_dag">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">filter_dag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">priority_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outbase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutability_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitution_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotypemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotypemap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotype_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">img_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'svg'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#filter_dag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.filter_dag" title="Permalink to this definition"></a></dt>
<dd><p>Filter trees in the provided history DAG.</p>
<p>Trim the provided history DAG, containing sequence labels, by-node
abundance attributes, and a <cite>parameters</cite> attribute, to minimize a linear
combination of branching process likelihood, isotype parsimony score,
mutability parsimony score, and number of alleles, with coefficients
provided in the argument <cite>priority_weights</cite>, in that order.</p>
<p>To compute each tree’s combined score, a linear transformation is applied to each trait,
so that the best observed value is mapped to 0, and the worst observed value is mapped to
1. The tree score is the weighted sum of transformed traits, using passed coefficients.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dag</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">HistoryDag</span></code>) – A history DAG object to trim</p></li>
<li><p><strong>priority_weights</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]]) – A sequence of coefficients for prioritizing tree weights.
The order of coefficients is: branching process likelihood, isotype
parsimony score, mutability parsimony score, and number of alleles.
If priority_weights is not provided, trees will be ranked lexicographically
by traits, in the same order.</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – print information about trimming</p></li>
<li><p><strong>outbase</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – file name stem for a file with information for each tree in the DAG, and rank plots of likelihoods.
If not provided, no such files will be written.</p></li>
<li><p><strong>mutability_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A mutability model</p></li>
<li><p><strong>substitution_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A substitution model</p></li>
<li><p><strong>isotypemap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A mapping of sequences to isotypes</p></li>
<li><p><strong>isotypemap</strong> – A dictionary mapping original IDs to observed isotype names</p></li>
<li><p><strong>isotypemap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>isotypemap</cite></p></li>
<li><p><strong>idmap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Set" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]]) – A dictionary mapping unique sequence IDs to sets of original IDs of observed sequences</p></li>
<li><p><strong>idmap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>idmap</cite></p></li>
<li><p><strong>isotype_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A sequence of isotype names containing values in <cite>isotypemap</cite>, in the correct switching order</p></li>
<li><p><strong>img_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – Format to be used for output plots, if <cite>outbase</cite> string is provided.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The trimmed history DAG, containing all optimal trees according to specified criteria.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.ll_cmcount_dagfuncs">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">ll_cmcount_dagfuncs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#ll_cmcount_dagfuncs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.ll_cmcount_dagfuncs" title="Permalink to this definition"></a></dt>
<dd><p>A slower but more numerically stable equivalent to ll_genotype_dagfuncs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching process parameters</p></li>
<li><p><strong>q</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching process parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">historydag.utils.AddFuncDict</span></code> object, which can be unpacked and used
as keyword arguments for DAG trimming and weight annotation according to
branching process likelihood.
Weight format is <code class="docutils literal notranslate"><span class="pre">(log</span> <span class="pre">likelihood,</span> <span class="pre">cmcounts)</span></code> where cmcounts is a FrozenMultiset containing abundance, mutant clade count pairs.
To use these functions for DAG trimming, use an optimal function like
<cite>lambda l: max(l, key=lambda ll: ll[0])</cite> for clarity, although min or max should work too.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.ll_genotype_dagfuncs">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">ll_genotype_dagfuncs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#ll_genotype_dagfuncs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.ll_genotype_dagfuncs" title="Permalink to this definition"></a></dt>
<dd><p>Return functions for counting tree log likelihood on the history DAG.</p>
<p>Although these functions are fast, for numerical consistency use
<a class="reference internal" href="gctree.dag_filtering.ll_cmcount_dagfuncs.html#gctree.dag_filtering.ll_cmcount_dagfuncs" title="gctree.dag_filtering.ll_cmcount_dagfuncs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ll_cmcount_dagfuncs()</span></code></a> instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching process parameters</p></li>
<li><p><strong>q</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching process parameters</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">historydag.utils.AddFuncDict</span></code> object, which can be unpacked and used
as keyword arguments for DAG trimming and weight annotation according to
branching process likelihood.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.mutability_dagfuncs">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">mutability_dagfuncs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#mutability_dagfuncs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.mutability_dagfuncs" title="Permalink to this definition"></a></dt>
<dd><p>Return functions for counting mutability parsimony on the history DAG.</p>
<p>Returns a <code class="docutils literal notranslate"><span class="pre">historydag.AddFuncDict</span></code> containing functions for counting
tree-wise sums of mutability distances in the history DAG.</p>
<p>This may not be stable numerically, but we expect every tree to have
a unique mutability parsimony score for non-degenerate mutability models, so
so this shouldn’t matter in practice.</p>
<p>Arguments are passed to <code class="xref py py-meth docutils literal notranslate"><span class="pre">MutationModel()</span></code> constructor.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.mutability_distance">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">mutability_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation_model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#mutability_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.mutability_distance" title="Permalink to this definition"></a></dt>
<dd><p>Returns a fast distance function based on passed mutation_model.</p>
<p>First, caches computed mutabilities for k-mers with k // 2 N’s on either end. This
is pretty fast for k=5, but the distance function should be created once
and reused.</p>
<p>The returned distance function sums <span class="math notranslate nohighlight">\(-log(mutability * p)\)</span> over all
sites which do match between its two sequence arguments, where <code class="docutils literal notranslate"><span class="pre">mutability</span></code>
is the mutation frequency of the k-mer surrounding the mutated base (in the
first sequence argument) and <code class="docutils literal notranslate"><span class="pre">p</span></code> is the transition probability to the new
base.</p>
<p>Note that, in particular, this function is not symmetric on its  arguments.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gctree.dag_filtering.isotype_dagfuncs">
<span class="sig-prename descclassname"><span class="pre">gctree.dag_filtering.</span></span><span class="sig-name descname"><span class="pre">isotype_dagfuncs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">isotypemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotypemap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotype_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/dag_filtering.html#isotype_dagfuncs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.dag_filtering.isotype_dagfuncs" title="Permalink to this definition"></a></dt>
<dd><p>Return functions for filtering by isotype parsimony score on the history
DAG.</p>
<p>Returns a <code class="docutils literal notranslate"><span class="pre">historydag.utils.AddFuncDict</span></code> which contains functions
necessary for filtering by isotype parsimony score. These functions return
weights which are a tuple containing a progressive isotype score, and a
frozenset containing isotypes used internally to compute isotype scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>isotypemap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A dictionary mapping original IDs to observed isotype names</p></li>
<li><p><strong>isotypemap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>isotypemap</cite></p></li>
<li><p><strong>idmap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Set" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]]) – A dictionary mapping unique sequence IDs to sets of original IDs of observed sequences</p></li>
<li><p><strong>idmap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>idmap</cite></p></li>
<li><p><strong>isotype_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A sequence of isotype names containing values in <cite>isotypemap</cite>, in the correct switching order</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <code class="docutils literal notranslate"><span class="pre">historydag.utils.AddFuncDict</span></code> which may be passed as keyword arguments
to <code class="docutils literal notranslate"><span class="pre">historydag.HistoryDag</span></code> methods which trim or annotate by weight.
These functions return weights which are a tuple containing
a progressive isotype score, and a frozenset containing isotypes used for computing isotype scores.</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="gctree.isotyping.IsotypeTemplate.html" class="btn btn-neutral float-left" title="gctree.isotyping.IsotypeTemplate" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="gctree.dag_filtering.filter_dag.html" class="btn btn-neutral float-right" title="gctree.dag_filtering.filter_dag" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, William DeWitt.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>