<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gctree.CollapsedForest &mdash; gctree  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="gctree.mutation_model" href="gctree.mutation_model.html" />
    <link rel="prev" title="gctree.CollapsedTree" href="gctree.CollapsedTree.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html">
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cite.html">Citing gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rendering-demo.html">Custom tree visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">CLI Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../cli.html">gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cli.html#parsimony-utilities">Parsimony utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cli.html#isotype">isotype</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Classes</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="gctree.CollapsedTree.html">gctree.CollapsedTree</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">gctree.CollapsedForest</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html#modules">Modules</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#developer-tools">Developer tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#todo-list">Todo list</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">gctree</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../api.html">Classes</a> &raquo;</li>
      <li>gctree.CollapsedForest</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="gctree-collapsedforest">
<h1>gctree.CollapsedForest<a class="headerlink" href="#gctree-collapsedforest" title="Permalink to this headline"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="gctree.CollapsedForest">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">gctree.</span></span><span class="sig-name descname"><span class="pre">CollapsedForest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A collection of trees.</p>
<p>We can intialize with a list of trees, each an instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">ete3.Tree</span></code> or <a class="reference internal" href="gctree.CollapsedTree.html#gctree.CollapsedTree" title="gctree.CollapsedTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">CollapsedTree</span></code></a>, or we can simulate the forest later.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="gctree.CollapsedForest.n_trees">
<span class="sig-name descname"><span class="pre">n_trees</span></span><a class="headerlink" href="#gctree.CollapsedForest.n_trees" title="Permalink to this definition"></a></dt>
<dd><p>number of trees in forest</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="gctree.CollapsedForest.parameters">
<span class="sig-name descname"><span class="pre">parameters</span></span><a class="headerlink" href="#gctree.CollapsedForest.parameters" title="Permalink to this definition"></a></dt>
<dd><p>fit branching process parameters, if mle has been run, otherwise None</p>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>forest</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference internal" href="gctree.branching_processes.CollapsedTree.html#gctree.branching_processes.CollapsedTree" title="gctree.branching_processes.CollapsedTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">CollapsedTree</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">TreeNode</span></code>]]]) – list of <code class="xref py py-class docutils literal notranslate"><span class="pre">ete3.Tree</span></code></p></li>
<li><p><strong>sequence_counts</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]]) – a mapping of observed sequences to observed abundances</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gctree.CollapsedForest.filter_trees" title="gctree.CollapsedForest.filter_trees"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_trees</span></code></a></p></td>
<td><p>Filter trees according to specified criteria.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gctree.CollapsedForest.iter_topology_classes" title="gctree.CollapsedForest.iter_topology_classes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_topology_classes</span></code></a></p></td>
<td><p>Sort trees by topology class</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gctree.CollapsedForest.likelihood_rankplot" title="gctree.CollapsedForest.likelihood_rankplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">likelihood_rankplot</span></code></a></p></td>
<td><p>save a rank plot of likelihoods to the file <cite>[outbase].inference.likelihood_rank.[img_type]</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gctree.CollapsedForest.ll" title="gctree.CollapsedForest.ll"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ll</span></code></a></p></td>
<td><p>Log likelihood of branching process parameters <span class="math notranslate nohighlight">\((p, q)\)</span> given tree topologies <span class="math notranslate nohighlight">\(T_1, \dots, T_n\)</span> and corresponding genotype abundances vectors <span class="math notranslate nohighlight">\(A_1, \dots, A_n\)</span> for each of <span class="math notranslate nohighlight">\(n\)</span> trees in the forest.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gctree.CollapsedForest.mle" title="gctree.CollapsedForest.mle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mle</span></code></a></p></td>
<td><p>Maximum likelihood estimate of <span class="math notranslate nohighlight">\((p, q)\)</span>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gctree.CollapsedForest.n_topologies" title="gctree.CollapsedForest.n_topologies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">n_topologies</span></code></a></p></td>
<td><p>Count the number of topology classes, ignoring internal node sequences</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gctree.CollapsedForest.sample_tree" title="gctree.CollapsedForest.sample_tree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_tree</span></code></a></p></td>
<td><p>Sample a random CollapsedTree from the forest</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gctree.CollapsedForest.simulate" title="gctree.CollapsedForest.simulate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">simulate</span></code></a></p></td>
<td><p>Simulate a forest of collapsed trees.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.simulate">
<span class="sig-name descname"><span class="pre">simulate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_trees</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.simulate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.simulate" title="Permalink to this definition"></a></dt>
<dd><p>Simulate a forest of collapsed trees. Overwrites existing forest attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching probability</p></li>
<li><p><strong>q</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – mutation probability</p></li>
<li><p><strong>n_trees</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – number of trees</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.ll">
<span class="sig-name descname"><span class="pre">ll</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marginal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.ll"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.ll" title="Permalink to this definition"></a></dt>
<dd><p>Log likelihood of branching process parameters <span class="math notranslate nohighlight">\((p, q)\)</span> given tree topologies <span class="math notranslate nohighlight">\(T_1, \dots, T_n\)</span> and corresponding genotype abundances vectors <span class="math notranslate nohighlight">\(A_1, \dots, A_n\)</span> for each of <span class="math notranslate nohighlight">\(n\)</span> trees in the forest.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">marginal=False</span></code> (the default), compute the joint log likelihood</p>
<div class="math notranslate nohighlight">
\[\ell(p, q; T, A) = \sum_{i=1}^n\log\mathbb{P}(T_i, A_i \mid p, q),\]</div>
<p>otherwise compute the marginal log likelihood</p>
<div class="math notranslate nohighlight">
\[\ell(p, q; T, A) = \log\left(\sum_{i=1}^n\mathbb{P}(T_i, A_i \mid p, q)\right).\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>p</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – branching probability</p></li>
<li><p><strong>q</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>) – mutation probability</p></li>
<li><p><strong>marginal</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – compute the marginal likelihood over trees, otherwise compute the joint likelihood of trees</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>, <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.22)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Log likelihood <span class="math notranslate nohighlight">\(\ell(p, q; T, A)\)</span> and its gradient <span class="math notranslate nohighlight">\(\nabla\ell(p, q; T, A)\)</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.mle">
<span class="sig-name descname"><span class="pre">mle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.mle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.mle" title="Permalink to this definition"></a></dt>
<dd><p>Maximum likelihood estimate of <span class="math notranslate nohighlight">\((p, q)\)</span>.</p>
<div class="math notranslate nohighlight">
\[(p, q) = \arg\max_{p,q\in [0,1]}\ell(p, q)\]</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>kwargs</strong> – keyword arguments passed along to the log likelihood <a class="reference internal" href="#gctree.CollapsedForest.ll" title="gctree.CollapsedForest.ll"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CollapsedForest.ll()</span></code></a></p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float64</span></code>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Tuple <span class="math notranslate nohighlight">\((p, q)\)</span> with estimated branching probability and estimated mutation probability</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.filter_trees">
<span class="sig-name descname"><span class="pre">filter_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ranking_coeffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutability_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitution_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotypemap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotypemap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idmap_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isotype_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outbase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gctree.out'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">summarize_forest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree_stats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.filter_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.filter_trees" title="Permalink to this definition"></a></dt>
<dd><p>Filter trees according to specified criteria.</p>
<p>Trim the forest to minimize a linear
combination of branching process likelihood, isotype parsimony score,
mutability parsimony score, and number of alleles, with coefficients
provided in the argument <a href="#id1"><span class="problematic" id="id2">``</span></a>ranking_coeffs`, in that order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ranking_coeffs</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>]]) – A list or tuple of coefficients for prioritizing tree weights.
The order of coefficients is: isotype parsimony score, mutability parsimony score,
and number of alleles. A coefficient of <code class="docutils literal notranslate"><span class="pre">-1</span></code> will be applied to branching process
likelihood.
If ranking_coeffs is not provided, trees will be ranked lexicographically
by likelihood, then by other traits, in the same order.</p></li>
<li><p><strong>mutability_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A mutability model</p></li>
<li><p><strong>substitution_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A substitution model</p></li>
<li><p><strong>isotypemap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A mapping of sequences to isotypes</p></li>
<li><p><strong>isotypemap</strong> – A dictionary mapping original IDs to observed isotype names</p></li>
<li><p><strong>isotypemap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>isotypemap</cite></p></li>
<li><p><strong>idmap</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Mapping</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Set" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]]) – A dictionary mapping unique sequence IDs to sets of original IDs of observed sequences</p></li>
<li><p><strong>idmap_file</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – A csv file providing an <cite>idmap</cite></p></li>
<li><p><strong>isotype_names</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.10)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – A sequence of isotype names containing values in <cite>isotypemap</cite>, in the correct switching order</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – print information about trimming</p></li>
<li><p><strong>outbase</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>) – file name stem for a file with information for each tree in the DAG.</p></li>
<li><p><strong>summarize_forest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – whether to write a summary of the forest to file <cite>[outbase].forest_summary.log</cite></p></li>
<li><p><strong>tree_stats</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – whether to write stats for each tree in the forest to file <cite>[outbase].tree_stats.log</cite></p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="gctree.branching_processes.CollapsedForest.html#gctree.branching_processes.CollapsedForest" title="gctree.branching_processes.CollapsedForest"><code class="xref py py-class docutils literal notranslate"><span class="pre">CollapsedForest</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The trimmed forest, containing all optimal trees according to the specified criteria, and a tuple
of data about the trees in that forest, with format (ll, isotype parsimony, mutability parsimony, alleles).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.likelihood_rankplot">
<span class="sig-name descname"><span class="pre">likelihood_rankplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outbase</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">img_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'svg'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.likelihood_rankplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.likelihood_rankplot" title="Permalink to this definition"></a></dt>
<dd><p>save a rank plot of likelihoods to the file <cite>[outbase].inference.likelihood_rank.[img_type]</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.n_topologies">
<span class="sig-name descname"><span class="pre">n_topologies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.n_topologies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.n_topologies" title="Permalink to this definition"></a></dt>
<dd><p>Count the number of topology classes, ignoring internal node sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.10)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.iter_topology_classes">
<span class="sig-name descname"><span class="pre">iter_topology_classes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.iter_topology_classes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.iter_topology_classes" title="Permalink to this definition"></a></dt>
<dd><p>Sort trees by topology class</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A generator of CollapsedForest objects, each containing trees with the same topology,</dt><dd><p>ignoring internal node labels. CollapsedForests will be yielded in reverse-order
of the number of trees in each topology class, so that each CollapsedForest will
contain at least as many trees as the one that follows.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gctree.CollapsedForest.sample_tree">
<span class="sig-name descname"><span class="pre">sample_tree</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/gctree/branching_processes.html#CollapsedForest.sample_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gctree.CollapsedForest.sample_tree" title="Permalink to this definition"></a></dt>
<dd><p>Sample a random CollapsedTree from the forest</p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="gctree.branching_processes.CollapsedTree.html#gctree.branching_processes.CollapsedTree" title="gctree.branching_processes.CollapsedTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">CollapsedTree</span></code></a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="gctree.CollapsedTree.html" class="btn btn-neutral float-left" title="gctree.CollapsedTree" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="gctree.mutation_model.html" class="btn btn-neutral float-right" title="gctree.mutation_model" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, William DeWitt.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>