<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gctree.branching_processes &mdash; gctree  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html">
            <img src="../../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cite.html">Citing gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../rendering-demo.html">Custom tree visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">CLI Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html#parsimony-utilities">Parsimony utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html#isotype">isotype</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html#modules">Modules</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html#developer-tools">Developer tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html#todo-list">Todo list</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gctree</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>gctree.branching_processes</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gctree.branching_processes</h1><div class="highlight"><pre>
<span></span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module contains classes for simulation and inference for a binary</span>
<span class="sd">branching process with mutation in which the tree is collapsed to nodes that</span>
<span class="sd">count the number of clonal leaves of each type.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">import</span> <span class="nn">gctree.utils</span>
<span class="kn">from</span> <span class="nn">gctree.isotyping</span> <span class="kn">import</span> <span class="n">_isotype_dagfuncs</span><span class="p">,</span> <span class="n">_isotype_annotation_dagfuncs</span>
<span class="kn">from</span> <span class="nn">gctree.mutation_model</span> <span class="kn">import</span> <span class="n">_mutability_dagfuncs</span>
<span class="kn">from</span> <span class="nn">gctree.phylip_parse</span> <span class="kn">import</span> <span class="n">disambiguate</span>

<span class="kn">from</span> <span class="nn">frozendict</span> <span class="kn">import</span> <span class="n">frozendict</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">scipy.special</span> <span class="k">as</span> <span class="nn">scs</span>
<span class="kn">import</span> <span class="nn">scipy.optimize</span> <span class="k">as</span> <span class="nn">sco</span>
<span class="kn">import</span> <span class="nn">ete3</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="kn">from</span> <span class="nn">Bio.Phylo.TreeConstruction</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">import</span> <span class="nn">collections</span> <span class="k">as</span> <span class="nn">coll</span>
<span class="kn">import</span> <span class="nn">historydag</span> <span class="k">as</span> <span class="nn">hdag</span>
<span class="kn">import</span> <span class="nn">multiset</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mp</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">Mapping</span><span class="p">,</span> <span class="n">Sequence</span>
<span class="kn">from</span> <span class="nn">decimal</span> <span class="kn">import</span> <span class="n">Decimal</span>

<span class="n">np</span><span class="o">.</span><span class="n">seterr</span><span class="p">(</span><span class="nb">all</span><span class="o">=</span><span class="s2">&quot;raise&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="CollapsedTree"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree">[docs]</a><span class="k">class</span> <span class="nc">CollapsedTree</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;A collapsed tree, modeled as an infinite type Galton-Watson process run</span>
<span class="sd">    to extinction.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        tree: :class:`ete3.TreeNode` object with ``abundance`` node features</span>

<span class="sd">    Args:</span>
<span class="sd">        tree: ete3 tree with ``abundance`` node features. Nonzero abundances expected on leaves,</span>
<span class="sd">            and optionally nodes adjacent to leaves via a path of length zero edges. If uncollapsed,</span>
<span class="sd">            it will be collapsed along branches with no mutations. Can be ommitted on initializaion, and later simulated.</span>
<span class="sd">            If a tree is provided, names of nodes with abundance 0 will not be preserved.</span>
<span class="sd">        allow_repeats: tolerate the existence of nodes with the same genotype after collapse, e.g. in sister clades.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_max_ll_cache</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

<div class="viewcode-block" id="CollapsedTree.__init__"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.html#gctree.CollapsedTree.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tree</span><span class="p">:</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TreeNode</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">allow_repeats</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">tree</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">def</span> <span class="nf">merge_isotype_dicts</span><span class="p">(</span><span class="n">parent_isotype</span><span class="p">,</span> <span class="n">child_isotype</span><span class="p">):</span>
                <span class="c1"># values are abundances and keys are isotypes.</span>
                <span class="n">parent_isotype</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">parent_isotype</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">child_isotype</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">parent_isotype</span><span class="p">:</span>
                        <span class="n">parent_isotype</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">parent_isotype</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="n">val</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">parent_isotype</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span>
                <span class="k">return</span> <span class="n">frozendict</span><span class="p">(</span><span class="n">parent_isotype</span><span class="p">)</span>

            <span class="c1"># remove unobserved internal unifurcations</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">prevent_nondicotomic</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

            <span class="c1"># ensure distances are correct before collapse</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">():</span>
                <span class="n">node</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="n">gctree</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">hamming_distance</span><span class="p">(</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">sequence</span>
                <span class="p">)</span>

            <span class="c1"># iterate over the tree below root and collapse edges of zero</span>
            <span class="c1"># length if the node is a leaf and it&#39;s parent has nonzero</span>
            <span class="c1"># abundance we combine taxa names to a set to acommodate</span>
            <span class="c1"># bootstrap samples that result in repeated genotypes</span>
            <span class="n">observed_genotypes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">((</span><span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">))</span>
            <span class="n">observed_genotypes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">get_descendants</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;postorder&quot;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">dist</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># if an abundance is nonzero, that&#39;s the right one.</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">abundance</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">abundance</span><span class="p">)</span>
                    <span class="c1"># isotype is dictionary with isotype as key and observed</span>
                    <span class="c1"># abundance as value</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">isotype</span> <span class="o">=</span> <span class="n">merge_isotype_dicts</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">isotype</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">isotype</span><span class="p">)</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">node_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">node_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">node_up_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">node_up_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">node_up_set</span> <span class="o">&lt;</span> <span class="n">observed_genotypes</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">node_set</span> <span class="o">&lt;</span> <span class="n">observed_genotypes</span><span class="p">:</span>
                            <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">node_set</span> <span class="o">|</span> <span class="n">node_up_set</span><span class="p">)</span>
                            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                                <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">node_set</span> <span class="o">&lt;</span> <span class="n">observed_genotypes</span><span class="p">:</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">node_set</span><span class="p">)</span>
                        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                            <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">prevent_nondicotomic</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="s2">&quot;inferred_isotype&quot;</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">isotype</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                <span class="s2">&quot;inferred_isotype&quot;</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">isotype</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
            <span class="p">)</span>

            <span class="n">final_observed_genotypes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">node</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
                    <span class="n">final_observed_genotypes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">final_observed_genotypes</span> <span class="o">!=</span> <span class="n">observed_genotypes</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;observed genotypes don&#39;t match after &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;collapse</span><span class="se">\n\t</span><span class="s2">before: </span><span class="si">{</span><span class="n">observed_genotypes</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n\t</span><span class="s2">after: </span><span class="si">{</span><span class="n">final_observed_genotypes</span><span class="si">}</span><span class="se">\n\t</span><span class="s2">&quot;</span>
                    <span class="s2">&quot;symmetric diff: &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">observed_genotypes</span> <span class="o">^</span> <span class="n">final_observed_genotypes</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="n">rep_seq</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">())</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">sequence</span>
                        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">allow_repeats</span> <span class="ow">and</span> <span class="n">rep_seq</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;Repeated observed sequences in collapsed &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;tree. </span><span class="si">{</span><span class="n">rep_seq</span><span class="si">}</span><span class="s2"> sequences were found repeated.&quot;</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">allow_repeats</span> <span class="ow">and</span> <span class="n">rep_seq</span><span class="p">:</span>
                <span class="n">rep_seq</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
                <span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="p">[</span>
                            <span class="n">node</span><span class="o">.</span><span class="n">sequence</span>
                            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
                            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="s2">&quot;Repeated observed sequences in collapsed tree. &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">rep_seq</span><span class="si">}</span><span class="s2"> sequences were found repeated.&quot;</span>
                <span class="p">)</span>
            <span class="c1"># a custom ladderize accounting for abundance and sequence to break</span>
            <span class="c1"># ties in abundance. In same traversal fix unobserved node names</span>
            <span class="n">unobserved_count</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">unobserved_dict</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;postorder&quot;</span><span class="p">):</span>
                <span class="c1"># add a partition feature and compute it recursively up tree</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="s2">&quot;partition&quot;</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">partition</span> <span class="k">for</span> <span class="n">node2</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># sort children of this node based on partion and sequence</span>
                <span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">node</span><span class="p">:</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">partition</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
                <span class="c1"># change node name if necessary</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">in</span> <span class="n">unobserved_dict</span><span class="p">:</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">unobserved_dict</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">unobserved_count</span><span class="p">)</span>
                        <span class="n">unobserved_count</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">unobserved_dict</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_build_cm_counts</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="kc">None</span></div>

    <span class="k">def</span> <span class="nf">_build_cm_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># create tuple (c, m) for each node, and store in a tuple of</span>
        <span class="c1"># ((c, m), n)&#39;s, where n is the multiplicity of (c, m) seen in the</span>
        <span class="c1"># tree, adding pseudocount to root if unobserved unifurcation at root.</span>
        <span class="n">cmlist</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="n">rootcm</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">abundance</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">rootcm</span> <span class="o">==</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">cmlist</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cmlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rootcm</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cm_counts</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">coll</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">cmlist</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_simulate_genotype</span><span class="p">(</span><span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate the number of clonal leaves :math:`c` and mutant clades</span>
<span class="sd">        :math:`m` as a Galton-Watson process with branching probability</span>
<span class="sd">        :math:`p` and mutation probability :math:`q`.</span>

<span class="sd">        Args:</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple :math:`(c, m)` of the number of clonal leaves and mutant clades</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">p</span> <span class="o">&lt;=</span> <span class="mi">1</span> <span class="ow">and</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">q</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;p and q must be in the unit interval: &quot;</span> <span class="sa">f</span><span class="s2">&quot;p = </span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">, q = </span><span class="si">{</span><span class="n">q</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">p</span> <span class="o">&gt;=</span> <span class="mf">0.5</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;p = </span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2"> is not subcritical, tree simulations not garanteed to terminate!&quot;</span>
            <span class="p">)</span>
        <span class="c1"># let&#39;s track the tree in breadth first order, listing number of clonal</span>
        <span class="c1"># and mutant descendants of each node mutant clades terminate in this</span>
        <span class="c1"># view</span>
        <span class="n">cumsum_clones</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">len_tree</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">c</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">m</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># while termination condition not met</span>
        <span class="k">while</span> <span class="n">cumsum_clones</span> <span class="o">&gt;</span> <span class="n">len_tree</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">p</span><span class="p">:</span>
                <span class="n">mutants</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">q</span> <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
                <span class="n">clones</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">-</span> <span class="n">mutants</span>
                <span class="n">m</span> <span class="o">+=</span> <span class="n">mutants</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">mutants</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">clones</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">c</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">cumsum_clones</span> <span class="o">+=</span> <span class="n">clones</span>
            <span class="n">len_tree</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">assert</span> <span class="n">cumsum_clones</span> <span class="o">==</span> <span class="n">len_tree</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">c</span><span class="p">,</span> <span class="n">m</span>

    <span class="nd">@staticmethod</span>
    <span class="nd">@functools</span><span class="o">.</span><span class="n">lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">_ll_genotype</span><span class="p">(</span>
        <span class="n">c</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Log-probability of getting :math:`c` leaves that are clones of the root and</span>
<span class="sd">        :math:`m` mutant clades off the root lineage, given branching probability :math:`p` and</span>
<span class="sd">        mutation probability :math:`q`.</span>

<span class="sd">        AKA the spaceship distribution. Also returns gradient wrt p and q</span>
<span class="sd">        (p, q). Computed by dynamic programming.</span>

<span class="sd">        Args:</span>
<span class="sd">            c: clonal leaves</span>
<span class="sd">            m: mutant clades</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>

<span class="sd">        Returns:</span>
<span class="sd">            log-likelihood and gradient wrt :math:`p` and :math:`q`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span> <span class="ow">in</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_max_ll_cache</span><span class="p">:</span>
            <span class="n">cached_c</span><span class="p">,</span> <span class="n">cached_m</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_max_ll_cache</span><span class="p">[(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># clear cache for old parameters if parameters change:</span>
            <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_max_ll_cache</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="o">.</span><span class="n">cache_clear</span><span class="p">()</span>
            <span class="n">cached_c</span><span class="p">,</span> <span class="n">cached_m</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
        <span class="c1"># Check cache is built</span>
        <span class="k">if</span> <span class="n">c</span> <span class="o">&gt;</span> <span class="n">cached_c</span> <span class="ow">or</span> <span class="n">m</span> <span class="o">&gt;</span> <span class="n">cached_m</span><span class="p">:</span>
            <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_max_ll_cache</span><span class="p">[(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)]</span> <span class="o">=</span> <span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">m</span><span class="p">)</span>
            <span class="c1"># Build missing cache in three parts:</span>
            <span class="c1"># |1 3</span>
            <span class="c1"># |X 2</span>
            <span class="c1"># Where &#39;X&#39; is already built, axes are c and m.</span>
            <span class="k">for</span> <span class="n">cx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">mx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_m</span><span class="p">,</span> <span class="n">m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">cx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mx</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">cx</span><span class="p">,</span> <span class="n">mx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">mx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">cx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_c</span><span class="p">,</span> <span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">cx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mx</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">cx</span><span class="p">,</span> <span class="n">mx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">mx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">cx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cached_c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">cx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mx</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">cx</span><span class="p">,</span> <span class="n">mx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
            <span class="c1"># If we&#39;re here, we&#39;ve computed what we want</span>
            <span class="k">return</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">c</span> <span class="o">==</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="p">(</span><span class="n">c</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Zero likelihood event&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">c</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logf_result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span>
            <span class="n">dlogfdp_result</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span>
            <span class="n">dlogfdq_result</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">elif</span> <span class="n">c</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">logf_result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">q</span><span class="p">)</span>
            <span class="n">dlogfdp_result</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">p</span>
            <span class="n">dlogfdq_result</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">/</span> <span class="n">q</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">m</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="p">(</span>
                    <span class="n">neighbor_ll_genotype</span><span class="p">,</span>
                    <span class="p">(</span>
                        <span class="n">neighbor_dlogfdp</span><span class="p">,</span>
                        <span class="n">neighbor_dlogfdq</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">m</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
                <span class="n">logg_array</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="p">(</span>
                        <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                        <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
                        <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">q</span><span class="p">)</span>
                        <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span>
                        <span class="o">+</span> <span class="n">neighbor_ll_genotype</span>
                    <span class="p">)</span>
                <span class="p">]</span>
                <span class="n">dloggdp_array</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="o">/</span> <span class="n">p</span> <span class="o">+</span> <span class="n">neighbor_dlogfdp</span><span class="p">]</span>
                <span class="n">dloggdq_array</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span> <span class="o">/</span> <span class="n">q</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span> <span class="o">+</span> <span class="n">neighbor_dlogfdq</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logg_array</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">dloggdp_array</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">dloggdq_array</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">cx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">c</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">mx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">cx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">mx</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">c</span> <span class="o">-</span> <span class="n">cx</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">m</span> <span class="o">-</span> <span class="n">mx</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">):</span>
                        <span class="p">(</span>
                            <span class="n">neighbor1_ll_genotype</span><span class="p">,</span>
                            <span class="p">(</span><span class="n">neighbor1_dlogfdp</span><span class="p">,</span> <span class="n">neighbor1_dlogfdq</span><span class="p">),</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">cx</span><span class="p">,</span> <span class="n">mx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
                        <span class="p">(</span>
                            <span class="n">neighbor2_ll_genotype</span><span class="p">,</span>
                            <span class="p">(</span><span class="n">neighbor2_dlogfdp</span><span class="p">,</span> <span class="n">neighbor2_dlogfdq</span><span class="p">),</span>
                        <span class="p">)</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">c</span> <span class="o">-</span> <span class="n">cx</span><span class="p">,</span> <span class="n">m</span> <span class="o">-</span> <span class="n">mx</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
                        <span class="n">logg</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
                            <span class="o">+</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span>
                            <span class="o">+</span> <span class="n">neighbor1_ll_genotype</span>
                            <span class="o">+</span> <span class="n">neighbor2_ll_genotype</span>
                        <span class="p">)</span>
                        <span class="n">dloggdp</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">p</span> <span class="o">+</span> <span class="n">neighbor1_dlogfdp</span> <span class="o">+</span> <span class="n">neighbor2_dlogfdp</span>
                        <span class="n">dloggdq</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span> <span class="o">+</span> <span class="n">neighbor1_dlogfdq</span> <span class="o">+</span> <span class="n">neighbor2_dlogfdq</span>
                        <span class="n">logg_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">logg</span><span class="p">)</span>
                        <span class="n">dloggdp_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dloggdp</span><span class="p">)</span>
                        <span class="n">dloggdq_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dloggdq</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">logg_array</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Zero likelihood event&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logf_result</span> <span class="o">=</span> <span class="n">scs</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">(</span><span class="n">logg_array</span><span class="p">)</span>
                <span class="n">softmax_logg_array</span> <span class="o">=</span> <span class="n">scs</span><span class="o">.</span><span class="n">softmax</span><span class="p">(</span><span class="n">logg_array</span><span class="p">)</span>
                <span class="n">dlogfdp_result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">softmax_logg_array</span><span class="p">,</span> <span class="n">dloggdp_array</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="n">dlogfdq_result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="n">softmax_logg_array</span><span class="p">,</span> <span class="n">dloggdq_array</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">logf_result</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">dlogfdp_result</span><span class="p">,</span> <span class="n">dlogfdq_result</span><span class="p">]))</span>

<div class="viewcode-block" id="CollapsedTree.ll"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.ll">[docs]</a>    <span class="k">def</span> <span class="nf">ll</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
        <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Log likelihood of branching process parameters :math:`(p, q)` given tree topology :math:`T` and genotype abundances :math:`A`.</span>

<span class="sd">        .. math::</span>
<span class="sd">            \ell(p, q; T, A) = \log\mathbb{P}(T, A \mid p, q)</span>

<span class="sd">        Args:</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>

<span class="sd">        Returns:</span>
<span class="sd">            Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;tree data must be defined to compute likelihood&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">_lltree</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_cm_counts</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedTree.mle"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.mle">[docs]</a>    <span class="k">def</span> <span class="nf">mle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Maximum likelihood estimate of :math:`(p, q)`.</span>

<span class="sd">        .. math::</span>
<span class="sd">            (p, q) = \arg\max_{p,q\in [0,1]}\ell(p, q)</span>

<span class="sd">        Args:</span>
<span class="sd">            kwargs: keyword arguments passed along to the log likelihood :meth:`CollapsedTree.ll`</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple :math:`(p, q)` with estimated branching probability and estimated mutation probability</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">_mle_helper</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ll</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedTree.simulate"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.simulate">[docs]</a>    <span class="k">def</span> <span class="nf">simulate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">root</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate a collapsed tree as an infinite type Galton-Watson process</span>
<span class="sd">        run to extintion, with branching probability :math:`p` and mutation</span>
<span class="sd">        probability :math:`q`. Overwrites existing tree attribute.</span>

<span class="sd">        Args:</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>
<span class="sd">            root: flag indicating simulation is being run from the root of the tree, so we should update tree attributes (should usually be ``True``)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">c</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_simulate_genotype</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tree</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TreeNode</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;abundance&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">):</span>
            <span class="c1"># ooooh, recursion</span>
            <span class="n">child</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="p">()</span>
            <span class="n">child</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">root</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">child</span> <span class="o">=</span> <span class="n">child</span><span class="o">.</span><span class="n">tree</span>
            <span class="n">child</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">root</span><span class="p">:</span>
            <span class="c1"># create list of (c, m) for each node</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_build_cm_counts</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;return a string representation for printing.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">)</span>

<div class="viewcode-block" id="CollapsedTree.render"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.render">[docs]</a>    <span class="k">def</span> <span class="nf">render</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">outfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">scale</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">branch_margin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">node_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">idlabel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">colormap</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">frame</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">position_map</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">chain_split</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">frame2</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">position_map2</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">show_support</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Render to tree image file.</span>

<span class="sd">        Args:</span>
<span class="sd">            outfile: file name to render to, filetype inferred from suffix, .svg for color</span>
<span class="sd">            scale: branch length scale in pixels (set automatically if ``None``)</span>
<span class="sd">            branch_margin: additional leaf branch separation margin, in pixels, to scale tree width</span>
<span class="sd">            node_size: size of nodes in pixels (set according to abundance if ``None``)</span>
<span class="sd">            idlabel: label nodes with seq ids, and write sequences of all nodes to a fasta file with same base name as ``outfile``</span>
<span class="sd">            colormap: dictionary mapping node names to color names or to dictionaries of color frequencies</span>
<span class="sd">            frame: coding frame for annotating amino acid substitutions</span>
<span class="sd">            position_map: mapping of position names for sequence indices, to be used with substitution annotations and the ``frame`` argument</span>
<span class="sd">            chain_split: if sequences are a concatenation two gene sequences, this is the index at which the 2nd one starts (requires ``frame`` and ``frame2`` arguments)</span>
<span class="sd">            frame2: coding frame for 2nd sequence when using ``chain_split``</span>
<span class="sd">            position_map2: like ``position_map``, but for 2nd sequence when using ``chain_split``</span>
<span class="sd">            show_support: annotate bootstrap support if available</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">frame</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">frame</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;frame must be 1, 2, or 3&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">my_layout</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">colormap</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">colormap</span><span class="p">:</span>
                <span class="n">circle_color</span> <span class="o">=</span> <span class="s2">&quot;lightgray&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">circle_color</span> <span class="o">=</span> <span class="n">colormap</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="n">text_color</span> <span class="o">=</span> <span class="s2">&quot;black&quot;</span>

            <span class="k">if</span> <span class="n">node_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">node_size2</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">))</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">circle_color</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">C</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">CircleFace</span><span class="p">(</span>
                        <span class="n">radius</span><span class="o">=</span><span class="n">node_size2</span><span class="p">,</span>
                        <span class="n">color</span><span class="o">=</span><span class="n">circle_color</span><span class="p">,</span>
                        <span class="n">label</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;text&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">),</span> <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">text_color</span><span class="p">}</span>
                        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span>
                        <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">C</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="mi">90</span>
                    <span class="n">C</span><span class="o">.</span><span class="n">hz_align</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">ete3</span><span class="o">.</span><span class="n">faces</span><span class="o">.</span><span class="n">add_face_to_node</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">P</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">PieChartFace</span><span class="p">(</span>
                        <span class="p">[</span><span class="mi">100</span> <span class="o">*</span> <span class="n">x</span> <span class="o">/</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">circle_color</span><span class="o">.</span><span class="n">values</span><span class="p">()],</span>
                        <span class="mi">2</span> <span class="o">*</span> <span class="n">node_size2</span><span class="p">,</span>
                        <span class="mi">2</span> <span class="o">*</span> <span class="n">node_size2</span><span class="p">,</span>
                        <span class="n">colors</span><span class="o">=</span><span class="p">[</span>
                            <span class="p">(</span><span class="n">color</span> <span class="k">if</span> <span class="n">color</span> <span class="o">!=</span> <span class="s2">&quot;None&quot;</span> <span class="k">else</span> <span class="s2">&quot;lightgray&quot;</span><span class="p">)</span>
                            <span class="k">for</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">circle_color</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
                        <span class="p">],</span>
                        <span class="n">line_color</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">T</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TextFace</span><span class="p">(</span>
                        <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">circle_color</span><span class="o">.</span><span class="n">values</span><span class="p">())]),</span>
                        <span class="n">tight_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">T</span><span class="o">.</span><span class="n">hz_align</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">T</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="mi">90</span>
                    <span class="n">ete3</span><span class="o">.</span><span class="n">faces</span><span class="o">.</span><span class="n">add_face_to_node</span><span class="p">(</span><span class="n">P</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="s2">&quot;branch-right&quot;</span><span class="p">)</span>
                    <span class="n">ete3</span><span class="o">.</span><span class="n">faces</span><span class="o">.</span><span class="n">add_face_to_node</span><span class="p">(</span><span class="n">T</span><span class="p">,</span> <span class="n">node</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="s2">&quot;branch-right&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">:</span>
                <span class="n">T</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TextFace</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">)</span>
                <span class="n">T</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="mi">90</span>
                <span class="n">T</span><span class="o">.</span><span class="n">hz_align</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">ete3</span><span class="o">.</span><span class="n">faces</span><span class="o">.</span><span class="n">add_face_to_node</span><span class="p">(</span>
                    <span class="n">T</span><span class="p">,</span>
                    <span class="n">node</span><span class="p">,</span>
                    <span class="mi">0</span><span class="p">,</span>
                    <span class="n">position</span><span class="o">=</span><span class="s2">&quot;branch-top&quot;</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">idlabel</span><span class="p">:</span>
                <span class="n">T</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TextFace</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">tight_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
                <span class="n">T</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="mi">90</span>
                <span class="n">T</span><span class="o">.</span><span class="n">hz_align</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">ete3</span><span class="o">.</span><span class="n">faces</span><span class="o">.</span><span class="n">add_face_to_node</span><span class="p">(</span>
                    <span class="n">T</span><span class="p">,</span>
                    <span class="n">node</span><span class="p">,</span>
                    <span class="mi">1</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">circle_color</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="mi">2</span><span class="p">,</span>
                    <span class="n">position</span><span class="o">=</span><span class="s2">&quot;branch-bottom&quot;</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="c1"># we render on a copy, so faces are not permanent</span>
        <span class="n">tree_copy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;deepcopy&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree_copy</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="n">nstyle</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">NodeStyle</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">colormap</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">colormap</span><span class="p">:</span>
                <span class="n">nstyle</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;lightgray&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">nstyle</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">colormap</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">node_size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">nstyle</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">node_size</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">nstyle</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;sequence&quot;</span> <span class="ow">in</span> <span class="n">tree_copy</span><span class="o">.</span><span class="n">features</span> <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">frame</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">chain_split</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">frame2</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                                <span class="s2">&quot;must define frame2 when using chain_split&quot;</span>
                            <span class="p">)</span>
                        <span class="k">if</span> <span class="n">frame2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chain_split</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                                <span class="s2">&quot;must define chain_split when using frame2&quot;</span>
                            <span class="p">)</span>
                        <span class="c1"># loop over split heavy/light chain subsequences</span>
                        <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">framex</span><span class="p">,</span> <span class="n">position_mapx</span> <span class="ow">in</span> <span class="p">(</span>
                            <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">chain_split</span><span class="p">,</span> <span class="n">frame</span><span class="p">,</span> <span class="n">position_map</span><span class="p">),</span>
                            <span class="p">(</span><span class="n">chain_split</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">frame2</span><span class="p">,</span> <span class="n">position_map2</span><span class="p">),</span>
                        <span class="p">):</span>
                            <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                <span class="n">seq</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
                                <span class="n">aa</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span>
                                    <span class="n">seq</span><span class="p">[</span>
                                        <span class="p">(</span><span class="n">framex</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="p">(</span>
                                            <span class="n">framex</span>
                                            <span class="o">-</span> <span class="mi">1</span>
                                            <span class="o">+</span> <span class="p">(</span><span class="mi">3</span> <span class="o">*</span> <span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">framex</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">//</span> <span class="mi">3</span><span class="p">))</span>
                                        <span class="p">)</span>
                                    <span class="p">]</span>
                                <span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
                                <span class="n">seq</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
                                <span class="n">aa_parent</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span>
                                    <span class="n">seq</span><span class="p">[</span>
                                        <span class="p">(</span><span class="n">framex</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="p">(</span>
                                            <span class="n">framex</span>
                                            <span class="o">-</span> <span class="mi">1</span>
                                            <span class="o">+</span> <span class="p">(</span><span class="mi">3</span> <span class="o">*</span> <span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">framex</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">//</span> <span class="mi">3</span><span class="p">))</span>
                                        <span class="p">)</span>
                                    <span class="p">]</span>
                                <span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
                                <span class="n">mutations</span> <span class="o">=</span> <span class="p">[</span>
                                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">aa1</span><span class="si">}{</span><span class="n">pos</span> <span class="k">if</span> <span class="n">position_mapx</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">position_mapx</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span><span class="si">}{</span><span class="n">aa2</span><span class="si">}</span><span class="s2">&quot;</span>
                                    <span class="k">for</span> <span class="n">pos</span><span class="p">,</span> <span class="p">(</span><span class="n">aa1</span><span class="p">,</span> <span class="n">aa2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">aa_parent</span><span class="p">,</span> <span class="n">aa</span><span class="p">))</span>
                                    <span class="k">if</span> <span class="n">aa1</span> <span class="o">!=</span> <span class="n">aa2</span>
                                <span class="p">]</span>
                                <span class="k">if</span> <span class="n">mutations</span><span class="p">:</span>
                                    <span class="n">T</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TextFace</span><span class="p">(</span>
                                        <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mutations</span><span class="p">),</span>
                                        <span class="n">fsize</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
                                        <span class="n">tight_text</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                        <span class="n">ftype</span><span class="o">=</span><span class="s2">&quot;Courier&quot;</span><span class="p">,</span>
                                    <span class="p">)</span>
                                    <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                                        <span class="n">T</span><span class="o">.</span><span class="n">margin_top</span> <span class="o">=</span> <span class="mi">6</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">T</span><span class="o">.</span><span class="n">margin_bottom</span> <span class="o">=</span> <span class="mi">6</span>
                                    <span class="n">T</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="o">-</span><span class="mi">90</span>
                                    <span class="n">node</span><span class="o">.</span><span class="n">add_face</span><span class="p">(</span>
                                        <span class="n">T</span><span class="p">,</span>
                                        <span class="mi">0</span><span class="p">,</span>
                                        <span class="n">position</span><span class="o">=</span><span class="s2">&quot;branch-bottom&quot;</span>
                                        <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="mi">0</span>
                                        <span class="k">else</span> <span class="s2">&quot;branch-top&quot;</span><span class="p">,</span>
                                    <span class="p">)</span>
                                <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">aa</span><span class="p">:</span>
                                    <span class="n">nstyle</span><span class="p">[</span><span class="s2">&quot;hz_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span>
            <span class="n">node</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">nstyle</span><span class="p">)</span>

        <span class="n">ts</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TreeStyle</span><span class="p">()</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">scale</span> <span class="o">=</span> <span class="n">scale</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">branch_vertical_margin</span> <span class="o">=</span> <span class="n">branch_margin</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">show_leaf_name</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">rotation</span> <span class="o">=</span> <span class="mi">90</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">draw_aligned_faces_as_table</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">allow_face_overlap</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">layout_fn</span> <span class="o">=</span> <span class="n">my_layout</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">show_scale</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">ts</span><span class="o">.</span><span class="n">show_branch_support</span> <span class="o">=</span> <span class="n">show_support</span>
        <span class="c1"># if we labelled seqs, let&#39;s also write the alignment out so we have</span>
        <span class="c1"># the sequences (including of internal nodes)</span>
        <span class="k">if</span> <span class="n">idlabel</span><span class="p">:</span>
            <span class="n">aln</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([])</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree_copy</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="n">aln</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">SeqRecord</span><span class="p">(</span>
                        <span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">)),</span>
                        <span class="nb">id</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">),</span>
                        <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;abundance=</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">aln</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">outfile</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">tree_copy</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedTree.feature_colormap"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.feature_colormap">[docs]</a>    <span class="k">def</span> <span class="nf">feature_colormap</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">feature</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">cmap</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;viridis&quot;</span><span class="p">,</span>
        <span class="n">vmin</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">vmax</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Generate a colormap based on a continuous tree feature.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature: feature name (all nodes in tree attribute must have this feature)</span>
<span class="sd">            cmap: any matplotlib color palette: https://matplotlib.org/stable/gallery/color/colormap_reference.html</span>
<span class="sd">            vmin: minimum value for colormap (default to minimum of the feature over the tree)</span>
<span class="sd">            vmax: maximum value for colormap (default to maximum of the feature over the tree)</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dictionary of node names to hex color strings, which may be used as the colormap in :meth:`gctree.CollapsedTree.render`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">vmin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">vmin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmin</span><span class="p">([</span><span class="nb">getattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()])</span>
        <span class="k">if</span> <span class="n">vmax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">vmax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">([</span><span class="nb">getattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()])</span>

        <span class="c1"># define the minimum and maximum values for our colormap</span>
        <span class="n">norm</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="n">vmin</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">vmax</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">{</span>
            <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">mp</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">to_hex</span><span class="p">(</span><span class="n">cmap</span><span class="p">(</span><span class="n">norm</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">feature</span><span class="p">))))</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
        <span class="p">}</span></div>

<div class="viewcode-block" id="CollapsedTree.write"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.write">[docs]</a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Serialize to pickle file.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_name: file name (.p suffix recommended)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedTree.newick"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.newick">[docs]</a>    <span class="k">def</span> <span class="nf">newick</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Write to newick file.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_name: file name (.nk suffix recommended)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="n">file_name</span><span class="p">,</span> <span class="n">format_root_node</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedTree.compare"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.compare">[docs]</a>    <span class="k">def</span> <span class="nf">compare</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">tree2</span><span class="p">:</span> <span class="n">CollapsedTree</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;identity&quot;</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="nb">bool</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Compare this tree to the other tree.</span>

<span class="sd">        Args:</span>
<span class="sd">            tree2: another object of this type</span>
<span class="sd">            method: comparison type (``identity``, ``MRCA``, or ``RF``)</span>

<span class="sd">        Returns:</span>
<span class="sd">            tree difference</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;identity&quot;</span><span class="p">:</span>
            <span class="c1"># we compare lists of seq, parent, abundance</span>
            <span class="c1"># return true if these lists are identical, else false</span>
            <span class="n">list1</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">sequence</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="n">list2</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">sequence</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree2</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="k">return</span> <span class="n">list1</span> <span class="o">==</span> <span class="n">list2</span>
        <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;MRCA&quot;</span><span class="p">:</span>
            <span class="c1"># matrix of hamming distance of common ancestors of taxa</span>
            <span class="c1"># takes a true and inferred tree as CollapsedTree objects</span>
            <span class="n">taxa</span> <span class="o">=</span> <span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span><span class="p">]</span>
            <span class="n">n_taxa</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">taxa</span><span class="p">)</span>
            <span class="n">d</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">n_taxa</span><span class="p">,</span> <span class="n">n_taxa</span><span class="p">))</span>
            <span class="n">sum_sites</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">n_taxa</span><span class="p">,</span> <span class="n">n_taxa</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_taxa</span><span class="p">):</span>
                <span class="n">nodei_true</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_search_nodes</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="n">taxa</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="fm">__next__</span><span class="p">()</span>
                <span class="n">nodei</span> <span class="o">=</span> <span class="n">tree2</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_search_nodes</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="n">taxa</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="fm">__next__</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_taxa</span><span class="p">):</span>
                    <span class="n">nodej_true</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_search_nodes</span><span class="p">(</span>
                        <span class="n">sequence</span><span class="o">=</span><span class="n">taxa</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                    <span class="p">)</span><span class="o">.</span><span class="fm">__next__</span><span class="p">()</span>
                    <span class="n">nodej</span> <span class="o">=</span> <span class="n">tree2</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_search_nodes</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="n">taxa</span><span class="p">[</span><span class="n">j</span><span class="p">])</span><span class="o">.</span><span class="fm">__next__</span><span class="p">()</span>
                    <span class="n">MRCA_true</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">nodei_true</span><span class="p">,</span> <span class="n">nodej_true</span><span class="p">)</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">sequence</span>
                    <span class="n">MRCA</span> <span class="o">=</span> <span class="n">tree2</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">((</span><span class="n">nodei</span><span class="p">,</span> <span class="n">nodej</span><span class="p">))</span><span class="o">.</span><span class="n">sequence</span>
                    <span class="n">d</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">gctree</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">hamming_distance</span><span class="p">(</span><span class="n">MRCA_true</span><span class="p">,</span> <span class="n">MRCA</span><span class="p">)</span>
                    <span class="n">sum_sites</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">MRCA_true</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">d</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">sum_sites</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;RF&quot;</span><span class="p">:</span>
            <span class="n">tree1_copy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;deepcopy&quot;</span><span class="p">)</span>
            <span class="n">tree2_copy</span> <span class="o">=</span> <span class="n">tree2</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;deepcopy&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">treex</span> <span class="ow">in</span> <span class="p">(</span><span class="n">tree1_copy</span><span class="p">,</span> <span class="n">tree2_copy</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">treex</span><span class="o">.</span><span class="n">traverse</span><span class="p">()):</span>
                    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">child</span> <span class="o">=</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TreeNode</span><span class="p">()</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
                        <span class="n">node</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">tree1_copy</span><span class="o">.</span><span class="n">robinson_foulds</span><span class="p">(</span>
                <span class="n">tree2_copy</span><span class="p">,</span>
                <span class="n">attr_t1</span><span class="o">=</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span>
                <span class="n">attr_t2</span><span class="o">=</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span>
                <span class="n">unrooted_trees</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;invalid distance method: &quot;</span> <span class="o">+</span> <span class="n">method</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="nf">_get_split</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node</span><span class="p">:</span> <span class="n">ete3</span><span class="o">.</span><span class="n">TreeNode</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Set</span><span class="p">,</span> <span class="n">Set</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Return the bipartition resulting from clipping this node&#39;s edge</span>
<span class="sd">        above.</span>

<span class="sd">        Args:</span>
<span class="sd">            node: tree node</span>

<span class="sd">        Returns:</span>
<span class="sd">            A tuple of two sets</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">get_tree_root</span><span class="p">()</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;node not found&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">node</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;this node is the root (no split above)&quot;</span><span class="p">)</span>
        <span class="n">parent</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span>
        <span class="n">taxa1</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">node2</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">node2</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">node2</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">taxa1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">taxa1</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">taxa1</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">taxa1</span><span class="p">)</span>
        <span class="n">node</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span>
        <span class="n">taxa2</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">node2</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">node2</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">node2</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">taxa2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">taxa2</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">taxa2</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">taxa2</span><span class="p">)</span>
        <span class="n">parent</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">taxa1</span><span class="o">.</span><span class="n">isdisjoint</span><span class="p">(</span><span class="n">taxa2</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">taxa1</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">taxa2</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">name</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">node</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span>
                <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">((</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">,)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">taxa1</span><span class="p">,</span> <span class="n">taxa2</span><span class="p">]))</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_split_compatibility</span><span class="p">(</span><span class="n">split1</span><span class="p">,</span> <span class="n">split2</span><span class="p">):</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">split1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">split1</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">^</span> <span class="n">split2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">split2</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">diff</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;splits do not cover the same taxa</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">taxa not in both: </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">partition1</span> <span class="ow">in</span> <span class="n">split1</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">partition2</span> <span class="ow">in</span> <span class="n">split2</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">partition1</span><span class="o">.</span><span class="n">isdisjoint</span><span class="p">(</span><span class="n">partition2</span><span class="p">):</span>
                    <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>

<div class="viewcode-block" id="CollapsedTree.support"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.support">[docs]</a>    <span class="k">def</span> <span class="nf">support</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">bootstrap_trees_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">CollapsedTree</span><span class="p">],</span>
        <span class="n">weights</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">compatibility</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Compute support from a list of bootstrap :class:`CollapsedTree` objects, and add to tree attibute.</span>

<span class="sd">        Args:</span>
<span class="sd">            bootstrap_trees_list: List of trees</span>
<span class="sd">            weights: weights for each tree, perhaps for weighting parsimony degenerate trees</span>
<span class="sd">            compatibility: counts trees that don&#39;t disconfirm the split.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">get_descendants</span><span class="p">():</span>
            <span class="n">split</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_split</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
            <span class="n">support</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">compatibility_</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tree</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bootstrap_trees_list</span><span class="p">):</span>
                <span class="n">compatible</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">supported</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">boot_node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">get_descendants</span><span class="p">():</span>
                    <span class="n">boot_split</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">_get_split</span><span class="p">(</span><span class="n">boot_node</span><span class="p">)</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">compatibility</span>
                        <span class="ow">and</span> <span class="n">compatible</span>
                        <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_split_compatibility</span><span class="p">(</span><span class="n">split</span><span class="p">,</span> <span class="n">boot_split</span><span class="p">)</span>
                    <span class="p">):</span>
                        <span class="n">compatible</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">compatibility</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">supported</span> <span class="ow">and</span> <span class="n">boot_split</span> <span class="o">==</span> <span class="n">split</span><span class="p">:</span>
                        <span class="n">supported</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">supported</span><span class="p">:</span>
                    <span class="n">support</span> <span class="o">+=</span> <span class="n">weights</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">weights</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">compatible</span><span class="p">:</span>
                    <span class="n">compatibility_</span> <span class="o">+=</span> <span class="n">weights</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">weights</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mi">1</span>
            <span class="n">node</span><span class="o">.</span><span class="n">support</span> <span class="o">=</span> <span class="n">compatibility_</span> <span class="k">if</span> <span class="n">compatibility</span> <span class="k">else</span> <span class="n">support</span></div>

<div class="viewcode-block" id="CollapsedTree.local_branching"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedTree.html#gctree.CollapsedTree.local_branching">[docs]</a>    <span class="k">def</span> <span class="nf">local_branching</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tau</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">tau0</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Add local branching statistics (Neher et al. 2014) as tree node</span>
<span class="sd">        features to the ETE tree attribute.</span>
<span class="sd">        After execution, all nodes will have new features ``LBI``</span>
<span class="sd">        (local branching index) and ``LBR`` (local branching ratio, below Vs</span>
<span class="sd">        above the node)</span>

<span class="sd">        Args:</span>
<span class="sd">            tau: decay timescale for exponential filter</span>
<span class="sd">            tau0: effective branch length for branches with zero mutations</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># the fixed integral contribution for clonal cells indicated by abundance annotations</span>
        <span class="n">clone_contribution</span> <span class="o">=</span> <span class="n">tau</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">tau0</span> <span class="o">/</span> <span class="n">tau</span><span class="p">))</span>

        <span class="c1"># post-order traversal to populate downward integral for each node</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;postorder&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="s2">&quot;LB_down&quot;</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">*</span> <span class="n">clone_contribution</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="mi">0</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="s2">&quot;LB_down&quot;</span><span class="p">,</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">*</span> <span class="n">clone_contribution</span>
                    <span class="o">+</span> <span class="nb">sum</span><span class="p">(</span>
                        <span class="n">tau</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">child</span><span class="o">.</span><span class="n">dist</span> <span class="o">/</span> <span class="n">tau</span><span class="p">))</span>
                        <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">child</span><span class="o">.</span><span class="n">dist</span> <span class="o">/</span> <span class="n">tau</span><span class="p">)</span> <span class="o">*</span> <span class="n">child</span><span class="o">.</span><span class="n">LB_down</span>
                        <span class="k">for</span> <span class="n">child</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">children</span>
                    <span class="p">),</span>
                <span class="p">)</span>

        <span class="c1"># pre-order traversal to populate upward integral for each node</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">(</span><span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;preorder&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">():</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;LB_up&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                    <span class="s2">&quot;LB_up&quot;</span><span class="p">,</span>
                    <span class="n">tau</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">node</span><span class="o">.</span><span class="n">dist</span> <span class="o">/</span> <span class="n">tau</span><span class="p">))</span>
                    <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">node</span><span class="o">.</span><span class="n">dist</span> <span class="o">/</span> <span class="n">tau</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">LB_up</span> <span class="o">+</span> <span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">LB_down</span><span class="p">),</span>
                <span class="p">)</span>

        <span class="c1"># finally, compute LBI (LBR) as the sum (ratio) of upward and downward integrals at each node</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;LBI&quot;</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">LB_down</span> <span class="o">+</span> <span class="n">node</span><span class="o">.</span><span class="n">LB_up</span><span class="p">)</span>
            <span class="n">node</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
                <span class="s2">&quot;LBR&quot;</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">LB_down</span> <span class="o">/</span> <span class="n">node</span><span class="o">.</span><span class="n">LB_up</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">()</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="p">)</span></div></div>


<span class="k">def</span> <span class="nf">_requires_dag</span><span class="p">(</span><span class="n">func</span><span class="p">):</span>
    <span class="nd">@functools</span><span class="o">.</span><span class="n">wraps</span><span class="p">(</span><span class="n">func</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">wrapper</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;CollapsedForest class method </span><span class="si">{</span><span class="n">func</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> may only be called on instances &quot;</span>
                <span class="s2">&quot;created from lists of ete3 trees. Instances created by simulation are not supported.&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">wrapper</span>


<div class="viewcode-block" id="CollapsedForest"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest">[docs]</a><span class="k">class</span> <span class="nc">CollapsedForest</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;A collection of trees.</span>

<span class="sd">    We can intialize with a list of trees, each an instance of :class:`ete3.Tree` or :class:`CollapsedTree`, or we can simulate the forest later.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        n_trees: number of trees in forest</span>
<span class="sd">        parameters: fit branching process parameters, if mle has been run, otherwise None</span>

<span class="sd">    Args:</span>
<span class="sd">        forest: list of :class:`ete3.Tree`</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="CollapsedForest.__init__"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.html#gctree.CollapsedForest.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">forest</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="n">CollapsedTree</span><span class="p">,</span> <span class="n">ete3</span><span class="o">.</span><span class="n">Tree</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="n">forest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">forest</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;passed empty tree list&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">forest</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">CollapsedTree</span><span class="p">):</span>
                <span class="n">forest</span> <span class="o">=</span> <span class="p">[</span><span class="n">ctree</span><span class="o">.</span><span class="n">tree</span> <span class="k">for</span> <span class="n">ctree</span> <span class="ow">in</span> <span class="n">forest</span><span class="p">]</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">forest</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ete3</span><span class="o">.</span><span class="n">Tree</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;If provided, `forest` argument should contain a list of ete3 trees or CollapsedTrees.&quot;</span>
                <span class="p">)</span>
            <span class="c1"># Collect stats for validation</span>
            <span class="n">model_tree</span> <span class="o">=</span> <span class="n">disambiguate</span><span class="p">(</span><span class="n">forest</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
            <span class="c1"># include root in counts, since deduplicate will never let observed</span>
            <span class="c1"># root be a leaf.</span>
            <span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">()}</span>
            <span class="n">counts</span><span class="p">[</span><span class="n">model_tree</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">abundance</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">,</span>
                <span class="s2">&quot;root&quot;</span><span class="p">:</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;parsimony_score&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">([</span><span class="n">node</span><span class="o">.</span><span class="n">dist</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">()]),</span>
                <span class="s2">&quot;root_seq&quot;</span><span class="p">:</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span>
                <span class="s2">&quot;leaf_seqs&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">model_tree</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()</span>
                <span class="p">},</span>
            <span class="p">}</span>
            <span class="c1"># Making this a private variable so that trying to access forest</span>
            <span class="c1"># attribute as before won&#39;t just give a confusing type or</span>
            <span class="c1"># attribute error.</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="o">=</span> <span class="n">_make_dag</span><span class="p">(</span><span class="n">forest</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">n_trees</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">count_trees</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">n_trees</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_isotyped</span> <span class="o">=</span> <span class="kc">False</span></div>

<div class="viewcode-block" id="CollapsedForest.simulate"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.simulate">[docs]</a>    <span class="k">def</span> <span class="nf">simulate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">n_trees</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate a forest of collapsed trees. Overwrites existing forest attribute.</span>

<span class="sd">        Args:</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>
<span class="sd">            n_trees: number of trees</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span> <span class="o">=</span> <span class="p">[</span><span class="n">CollapsedTree</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_trees</span><span class="p">)]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_trees</span> <span class="o">=</span> <span class="n">n_trees</span>
        <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span><span class="p">:</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
            <span class="n">coll</span><span class="o">.</span><span class="n">Counter</span><span class="p">([</span><span class="n">tree</span><span class="o">.</span><span class="n">_cm_counts</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span><span class="p">])</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.ll"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.ll">[docs]</a>    <span class="k">def</span> <span class="nf">ll</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
        <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
        <span class="n">marginal</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Log likelihood of branching process parameters :math:`(p, q)` given tree topologies :math:`T_1, \dots, T_n` and corresponding genotype abundances vectors :math:`A_1, \dots, A_n` for each of :math:`n` trees in the forest.</span>

<span class="sd">        If ``marginal=False`` (the default), compute the joint log likelihood</span>

<span class="sd">        .. math::</span>
<span class="sd">            \ell(p, q; T, A) = \sum_{i=1}^n\log\mathbb{P}(T_i, A_i \mid p, q),</span>

<span class="sd">        otherwise compute the marginal log likelihood</span>

<span class="sd">        .. math::</span>
<span class="sd">            \ell(p, q; T, A) = \log\left(\sum_{i=1}^n\mathbb{P}(T_i, A_i \mid p, q)\right).</span>

<span class="sd">        Args:</span>
<span class="sd">            p: branching probability</span>
<span class="sd">            q: mutation probability</span>
<span class="sd">            marginal: compute the marginal likelihood over trees, otherwise compute the joint likelihood of trees</span>

<span class="sd">        Returns:</span>
<span class="sd">            Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">cmcount_dagfuncs</span> <span class="o">=</span> <span class="n">_cmcounter_dagfuncs</span><span class="p">()</span>

                <span class="k">def</span> <span class="nf">to_tuple</span><span class="p">(</span><span class="n">mset</span><span class="p">):</span>
                    <span class="c1"># When there&#39;s unobserved root unifurcation, augment with</span>
                    <span class="c1"># pseudocount.</span>
                    <span class="k">if</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mset</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="n">mset</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)]</span> <span class="o">==</span> <span class="mi">1</span>
                        <span class="n">mset</span> <span class="o">=</span> <span class="n">mset</span> <span class="o">-</span> <span class="p">{(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)}</span> <span class="o">+</span> <span class="p">{(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)}</span>
                    <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">mset</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>

                <span class="n">cmcounters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">weight_count</span><span class="p">(</span><span class="o">**</span><span class="n">cmcount_dagfuncs</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="p">(</span><span class="n">to_tuple</span><span class="p">(</span><span class="n">mset</span><span class="p">),</span> <span class="n">mult</span><span class="p">)</span> <span class="k">for</span> <span class="n">mset</span><span class="p">,</span> <span class="n">mult</span> <span class="ow">in</span> <span class="n">cmcounters</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="p">]</span>
                <span class="c1"># an iterable containing tuples `(cm_counts, mult)`, where `cm_counts`</span>
                <span class="c1"># is an iterable describing a tree, and `mult` is the number of trees in the forest</span>
                <span class="c1"># which can be described with the iterable `cm_counts`.</span>
            <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
                    <span class="n">coll</span><span class="o">.</span><span class="n">Counter</span><span class="p">([</span><span class="n">tree</span><span class="o">.</span><span class="n">_cm_counts</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span><span class="p">])</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;forest data must be defined to compute likelihood&quot;</span><span class="p">)</span>

        <span class="n">terms</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="n">_lltree</span><span class="p">(</span><span class="n">cmcounts</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">),</span> <span class="n">count</span><span class="p">]</span> <span class="k">for</span> <span class="n">cmcounts</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cm_countlist</span>
        <span class="p">]</span>
        <span class="n">ls</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">term</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">terms</span><span class="p">])</span>
        <span class="n">grad_ls</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">term</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">terms</span><span class="p">])</span>
        <span class="c1"># This can be done ahead of time</span>
        <span class="n">count_ls</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">term</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">terms</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">marginal</span><span class="p">:</span>
            <span class="c1"># we need to find the smallest derivative component for each</span>
            <span class="c1"># coordinate, then subtract off to get positive things to logsumexp</span>
            <span class="n">grad_l</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">((</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">))):</span>
                <span class="n">i_prime</span> <span class="o">=</span> <span class="n">grad_ls</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">argmin</span><span class="p">()</span>
                <span class="n">b</span> <span class="o">=</span> <span class="p">(</span><span class="n">grad_ls</span><span class="p">[:,</span> <span class="n">j</span><span class="p">]</span> <span class="o">-</span> <span class="n">grad_ls</span><span class="p">[</span><span class="n">i_prime</span><span class="p">,</span> <span class="n">j</span><span class="p">])</span> <span class="o">*</span> <span class="n">count_ls</span>
                <span class="c1"># believe it or not, logsumexp can&#39;t handle 0 in b</span>
                <span class="c1"># when np.seterr(underflow=&#39;raise&#39;) on newer processors:</span>
                <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">b</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
                    <span class="c1"># logsumexp can&#39;t take empty list, but if b is 0, then</span>
                    <span class="c1"># logsumexp should be -inf, whose exponent is 0:</span>
                    <span class="n">res</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">res</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span>
                        <span class="n">scs</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">((</span><span class="n">ls</span> <span class="o">-</span> <span class="n">ls</span><span class="p">[</span><span class="n">i_prime</span><span class="p">])[</span><span class="n">b</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">],</span> <span class="n">b</span><span class="o">=</span><span class="n">b</span><span class="p">[</span><span class="n">b</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">])</span>
                        <span class="o">-</span> <span class="n">scs</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">(</span><span class="n">ls</span> <span class="o">-</span> <span class="n">ls</span><span class="p">[</span><span class="n">i_prime</span><span class="p">],</span> <span class="n">b</span><span class="o">=</span><span class="n">count_ls</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="n">grad_l</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">grad_ls</span><span class="p">[</span><span class="n">i_prime</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="n">res</span><span class="p">)</span>
            <span class="c1"># count_ls shouldn&#39;t have any zeros in it...</span>
            <span class="k">return</span> <span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">count_ls</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="o">+</span> <span class="n">scs</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">(</span><span class="n">ls</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="n">count_ls</span><span class="p">)),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="n">grad_l</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">ls</span> <span class="o">*</span> <span class="n">count_ls</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[(</span><span class="n">grad_ls</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">count_ls</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span> <span class="p">(</span><span class="n">grad_ls</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">count_ls</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()]</span>
            <span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.mle"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.mle">[docs]</a>    <span class="k">def</span> <span class="nf">mle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Maximum likelihood estimate of :math:`(p, q)`.</span>

<span class="sd">        .. math::</span>
<span class="sd">            (p, q) = \arg\max_{p,q\in [0,1]}\ell(p, q)</span>

<span class="sd">        Args:</span>
<span class="sd">            kwargs: keyword arguments passed along to the log likelihood :meth:`CollapsedForest.ll`</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple :math:`(p, q)` with estimated branching probability and estimated mutation probability</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span> <span class="o">=</span> <span class="n">_mle_helper</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ll</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span></div>

<div class="viewcode-block" id="CollapsedForest.filter_trees"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.filter_trees">[docs]</a>    <span class="nd">@_requires_dag</span>
    <span class="k">def</span> <span class="nf">filter_trees</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">ranking_coeffs</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">mutability_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">substitution_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ignore_isotype</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">chain_split</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">outbase</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;gctree.out&quot;</span><span class="p">,</span>
        <span class="n">summarize_forest</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">tree_stats</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">CollapsedForest</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Filter trees according to specified criteria.</span>

<span class="sd">        Trim the forest to minimize a linear</span>
<span class="sd">        combination of branching process likelihood, isotype parsimony score,</span>
<span class="sd">        mutability parsimony score, and number of alleles, with coefficients</span>
<span class="sd">        provided in the argument ``ranking_coeffs`, in that order.</span>

<span class="sd">        Args:</span>
<span class="sd">            ranking_coeffs: A list or tuple of coefficients for prioritizing tree weights.</span>
<span class="sd">                The order of coefficients is: isotype parsimony score, mutability parsimony score,</span>
<span class="sd">                and number of alleles. A coefficient of ``-1`` will be applied to branching process</span>
<span class="sd">                likelihood.</span>
<span class="sd">                If ranking_coeffs is not provided, trees will be ranked lexicographically</span>
<span class="sd">                by likelihood, then by other traits, in the same order.</span>
<span class="sd">            mutability_file: A mutability model</span>
<span class="sd">            substitution_file: A substitution model</span>
<span class="sd">            ignore_isotype: Ignore isotype parsimony when ranking. By default, isotype information added with</span>
<span class="sd">                :meth:``add_isotypes`` will be used to compute isotype parsimony, which is used in ranking.</span>
<span class="sd">            chain_split: The index at which non-adjacent sequences are concatenated, for calculating</span>
<span class="sd">                mutability parsimony.</span>
<span class="sd">            verbose: print information about trimming</span>
<span class="sd">            outbase: file name stem for a file with information for each tree in the DAG.</span>
<span class="sd">            summarize_forest: whether to write a summary of the forest to file `[outbase].forest_summary.log`</span>
<span class="sd">            tree_stats: whether to write stats for each tree in the forest to file `[outbase].tree_stats.log`</span>

<span class="sd">        Returns:</span>
<span class="sd">            The trimmed forest, containing all optimal trees according to the specified criteria, and a tuple</span>
<span class="sd">            of data about the trees in that forest, with format (ll, isotype parsimony, mutability parsimony, alleles).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">dag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mle</span><span class="p">(</span><span class="n">marginal</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">p</span><span class="p">,</span> <span class="n">q</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span>
        <span class="n">ll_dagfuncs</span> <span class="o">=</span> <span class="n">_ll_genotype_dagfuncs</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
        <span class="n">placeholder_dagfuncs</span> <span class="o">=</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;start_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;edge_weight_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                <span class="s2">&quot;accum_func&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="n">name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">ignore_isotype</span> <span class="ow">or</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_isotyped</span><span class="p">:</span>
            <span class="n">iso_funcs</span> <span class="o">=</span> <span class="n">placeholder_dagfuncs</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Isotype parsimony will be used as a ranking criterion&quot;</span><span class="p">)</span>
            <span class="c1"># Check for missing isotype data in all but root node, and fake root-adjacent leaf node</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span>
                <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;isotype&quot;</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">preorder</span><span class="p">()</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">()</span> <span class="ow">and</span> <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
            <span class="p">):</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                    <span class="s2">&quot;Some isotype data seems to be missing. Isotype parsimony scores may be incorrect.&quot;</span>
                <span class="p">)</span>

            <span class="n">iso_funcs</span> <span class="o">=</span> <span class="n">_isotype_dagfuncs</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">mutability_file</span> <span class="ow">and</span> <span class="n">substitution_file</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mutation model parsimony will be used as a ranking criterion&quot;</span><span class="p">)</span>

            <span class="n">mut_funcs</span> <span class="o">=</span> <span class="n">_mutability_dagfuncs</span><span class="p">(</span>
                <span class="n">mutability_file</span><span class="o">=</span><span class="n">mutability_file</span><span class="p">,</span>
                <span class="n">substitution_file</span><span class="o">=</span><span class="n">substitution_file</span><span class="p">,</span>
                <span class="n">splits</span><span class="o">=</span><span class="p">[]</span> <span class="k">if</span> <span class="n">chain_split</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="p">[</span><span class="n">chain_split</span><span class="p">],</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mut_funcs</span> <span class="o">=</span> <span class="n">placeholder_dagfuncs</span>
        <span class="n">allele_funcs</span> <span class="o">=</span> <span class="n">_allele_dagfuncs</span><span class="p">()</span>
        <span class="n">kwargls</span> <span class="o">=</span> <span class="p">(</span><span class="n">ll_dagfuncs</span><span class="p">,</span> <span class="n">iso_funcs</span><span class="p">,</span> <span class="n">mut_funcs</span><span class="p">,</span> <span class="n">allele_funcs</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ranking_coeffs</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ranking_coeffs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;If ranking_coeffs are provided to `filter_trees` method, a list of three values is expected.&quot;</span>
                <span class="p">)</span>
            <span class="n">coeffs</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">ranking_coeffs</span><span class="p">)</span>

            <span class="k">def</span> <span class="nf">minfunckey</span><span class="p">(</span><span class="n">weighttuple</span><span class="p">):</span>
                <span class="sd">&quot;&quot;&quot;Weighttuple will have (ll, isotypepars, mutabilitypars,</span>
<span class="sd">                alleles)&quot;&quot;&quot;</span>
                <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">priority</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="n">weight</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">priority</span><span class="p">,</span> <span class="n">weight</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">coeffs</span><span class="p">,</span> <span class="n">weighttuple</span><span class="p">)</span>
                    <span class="p">]</span>
                <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">def</span> <span class="nf">minfunckey</span><span class="p">(</span><span class="n">weighttuple</span><span class="p">):</span>
                <span class="sd">&quot;&quot;&quot;Weighttuple will have (ll, isotypepars, mutabilitypars,</span>
<span class="sd">                alleles)&quot;&quot;&quot;</span>
                <span class="c1"># Sort output by likelihood, then isotype parsimony, then mutability score</span>
                <span class="k">return</span> <span class="p">(</span><span class="o">-</span><span class="n">weighttuple</span><span class="p">[</span><span class="mi">0</span><span class="p">],)</span> <span class="o">+</span> <span class="n">weighttuple</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">def</span> <span class="nf">print_stats</span><span class="p">(</span><span class="n">statlist</span><span class="p">,</span> <span class="n">title</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">suppress_score</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
            <span class="n">show_score</span> <span class="o">=</span> <span class="n">ranking_coeffs</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">suppress_score</span>

            <span class="k">def</span> <span class="nf">reformat</span><span class="p">(</span><span class="n">field</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">field</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                    <span class="k">return</span> <span class="nb">format</span><span class="p">(</span><span class="n">field</span><span class="p">,</span> <span class="s2">&quot;&lt;&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">:{</span><span class="n">n</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">n</span><span class="si">}}</span><span class="s2">&quot;</span>

            <span class="k">def</span> <span class="nf">mask</span><span class="p">(</span><span class="n">weighttuple</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
                <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span>
                    <span class="n">reformat</span><span class="p">(</span><span class="n">field</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">kwargs</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">weighttuple</span><span class="p">,</span> <span class="n">kwargls</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">name</span>
                <span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parameters: </span><span class="si">{</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">title</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">)</span>
            <span class="n">statstring</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mask</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">kwargs</span> <span class="ow">in</span> <span class="n">kwargls</span><span class="p">),</span> <span class="n">n</span><span class="o">=</span><span class="mi">14</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;tree     </span><span class="se">\t</span><span class="si">{</span><span class="n">statstring</span><span class="si">}</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">treescore&quot;</span> <span class="k">if</span> <span class="n">show_score</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
                <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">best_weighttuple</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">statlist</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">statstring</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mask</span><span class="p">(</span><span class="n">best_weighttuple</span><span class="p">))</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">j</span><span class="si">:</span><span class="s2">&lt;10</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">statstring</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="o">+</span> <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">reformat</span><span class="p">(</span><span class="n">minfunckey</span><span class="p">(</span><span class="n">best_weighttuple</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="k">if</span> <span class="n">show_score</span>
                        <span class="k">else</span> <span class="s2">&quot;&quot;</span>
                    <span class="p">),</span>
                    <span class="n">file</span><span class="o">=</span><span class="n">file</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="c1"># Filter by likelihood, isotype parsimony, mutability,</span>
        <span class="c1"># and make ctrees, cforest, and render trees</span>
        <span class="n">dagweight_kwargs</span> <span class="o">=</span> <span class="n">ll_dagfuncs</span> <span class="o">+</span> <span class="n">iso_funcs</span> <span class="o">+</span> <span class="n">mut_funcs</span> <span class="o">+</span> <span class="n">allele_funcs</span>
        <span class="n">trimdag</span> <span class="o">=</span> <span class="n">dag</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">trimdag</span><span class="o">.</span><span class="n">trim_optimal_weight</span><span class="p">(</span>
            <span class="o">**</span><span class="n">dagweight_kwargs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">l</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">minfunckey</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># make sure trimming worked as expected:</span>
        <span class="n">min_weightcounter</span> <span class="o">=</span> <span class="n">trimdag</span><span class="o">.</span><span class="n">weight_count</span><span class="p">(</span><span class="o">**</span><span class="n">dagweight_kwargs</span><span class="p">)</span>
        <span class="n">min_weightset</span> <span class="o">=</span> <span class="p">{</span><span class="n">minfunckey</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">min_weightcounter</span><span class="p">}</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">min_weightset</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;Filtering was not successful. After trimming, these weights are represented:&quot;</span><span class="p">,</span>
                <span class="n">min_weightset</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">best_weighttuple</span> <span class="o">=</span> <span class="n">trimdag</span><span class="o">.</span><span class="n">optimal_weight_annotate</span><span class="p">(</span>
            <span class="o">**</span><span class="n">dagweight_kwargs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">l</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">minfunckey</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">summarize_forest</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outbase</span> <span class="o">+</span> <span class="s2">&quot;.forest_summary.log&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
                <span class="n">independent_best</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">kwargs</span> <span class="ow">in</span> <span class="n">kwargls</span><span class="p">:</span>
                    <span class="c1"># Only summarize for stats for which information was</span>
                    <span class="c1"># provided (not just placeholders):</span>
                    <span class="k">if</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                        <span class="n">independent_best</span><span class="o">.</span><span class="n">append</span><span class="p">([])</span>
                        <span class="k">for</span> <span class="n">opt</span> <span class="ow">in</span> <span class="p">[</span><span class="nb">min</span><span class="p">,</span> <span class="nb">max</span><span class="p">]:</span>
                            <span class="n">tempdag</span> <span class="o">=</span> <span class="n">dag</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                            <span class="n">opt_weight</span> <span class="o">=</span> <span class="n">tempdag</span><span class="o">.</span><span class="n">trim_optimal_weight</span><span class="p">(</span>
                                <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="n">opt</span>
                            <span class="p">)</span>
                            <span class="n">independent_best</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">opt_weight</span><span class="p">)</span>
                            <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                                <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Among trees with </span><span class="si">{</span><span class="n">opt</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">kwargs</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> of: </span><span class="si">{</span><span class="n">opt_weight</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
                            <span class="p">)</span>
                            <span class="k">for</span> <span class="n">inkwargs</span> <span class="ow">in</span> <span class="n">kwargls</span><span class="p">:</span>
                                <span class="k">if</span> <span class="n">inkwargs</span> <span class="o">!=</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="n">inkwargs</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                                    <span class="n">minval</span> <span class="o">=</span> <span class="n">tempdag</span><span class="o">.</span><span class="n">optimal_weight_annotate</span><span class="p">(</span>
                                        <span class="o">**</span><span class="n">inkwargs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="nb">min</span>
                                    <span class="p">)</span>
                                    <span class="n">maxval</span> <span class="o">=</span> <span class="n">tempdag</span><span class="o">.</span><span class="n">optimal_weight_annotate</span><span class="p">(</span>
                                        <span class="o">**</span><span class="n">inkwargs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="nb">max</span>
                                    <span class="p">)</span>
                                    <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                                        <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">inkwargs</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> range: </span><span class="si">{</span><span class="n">minval</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">maxval</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
                                    <span class="p">)</span>
                <span class="n">independent_best</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">reverse</span><span class="p">()</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">fh</span><span class="p">)</span>
                <span class="n">print_stats</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="p">[</span>
                            <span class="n">stat</span> <span class="o">-</span> <span class="n">best</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                            <span class="k">for</span> <span class="n">stat</span><span class="p">,</span> <span class="n">best</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">best_weighttuple</span><span class="p">,</span> <span class="n">independent_best</span><span class="p">)</span>
                        <span class="p">]</span>
                    <span class="p">],</span>
                    <span class="s2">&quot;Highest ranked tree: loss from best value&quot;</span><span class="p">,</span>
                    <span class="n">file</span><span class="o">=</span><span class="n">fh</span><span class="p">,</span>
                    <span class="n">suppress_score</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">tree_stats</span><span class="p">:</span>
            <span class="n">dag_ls</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">dag</span><span class="o">.</span><span class="n">weight_count</span><span class="p">(</span><span class="o">**</span><span class="n">dagweight_kwargs</span><span class="p">)</span><span class="o">.</span><span class="n">elements</span><span class="p">())</span>
            <span class="c1"># To clear _dp_data fields of their large cargo</span>
            <span class="n">dag</span><span class="o">.</span><span class="n">optimal_weight_annotate</span><span class="p">(</span><span class="n">edge_weight_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">:</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">dag_ls</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">minfunckey</span><span class="p">)</span>

            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dag_ls</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">dagweight_kwargs</span><span class="o">.</span><span class="n">names</span><span class="p">)</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outbase</span> <span class="o">+</span> <span class="s2">&quot;.tree_stats.csv&quot;</span><span class="p">)</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;set&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;all_trees&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
            <span class="n">bestdf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">best_weighttuple</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="n">dagweight_kwargs</span><span class="o">.</span><span class="n">names</span><span class="p">)</span>
            <span class="n">bestdf</span><span class="p">[</span><span class="s2">&quot;set&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;best_tree&quot;</span><span class="p">]</span>
            <span class="n">toplot_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span> <span class="n">bestdf</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">pplot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span>
                <span class="n">toplot_df</span><span class="p">[[</span><span class="s2">&quot;Log Likelihood&quot;</span><span class="p">,</span> <span class="s2">&quot;Isotype Pars.&quot;</span><span class="p">,</span> <span class="s2">&quot;Mut. Pars.&quot;</span><span class="p">,</span> <span class="s2">&quot;set&quot;</span><span class="p">]],</span>
                <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;set&quot;</span><span class="p">,</span>
                <span class="n">diag_kind</span><span class="o">=</span><span class="s2">&quot;hist&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">pplot</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">outbase</span> <span class="o">+</span> <span class="s2">&quot;.tree_stats.pairplot.png&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="n">print_stats</span><span class="p">([</span><span class="n">best_weighttuple</span><span class="p">],</span> <span class="s2">&quot;Stats for optimal trees&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_trimmed_self</span><span class="p">(</span><span class="n">trimdag</span><span class="p">),</span> <span class="n">best_weighttuple</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.likelihood_rankplot"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.likelihood_rankplot">[docs]</a>    <span class="k">def</span> <span class="nf">likelihood_rankplot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">outbase</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">img_type</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;save a rank plot of likelihoods to the file</span>
<span class="sd">        `[outbase].inference.likelihood_rank.[img_type]`.&quot;&quot;&quot;</span>
        <span class="n">ll_dagfuncs</span> <span class="o">=</span> <span class="n">_ll_genotype_dagfuncs</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">dag_l</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">ll</span><span class="p">)</span> <span class="k">for</span> <span class="n">ll</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">weight_count</span><span class="p">(</span><span class="o">**</span><span class="n">ll_dagfuncs</span><span class="p">)</span><span class="o">.</span><span class="n">elements</span><span class="p">()</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dag_l</span> <span class="o">=</span> <span class="p">[</span><span class="n">ctree</span><span class="o">.</span><span class="n">ll</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">ctree</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">]</span>
        <span class="n">dag_l</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="o">-</span><span class="n">n</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">6.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">dag_l</span><span class="p">),</span> <span class="s2">&quot;ko&quot;</span><span class="p">,</span> <span class="n">clip_on</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;gctree likelihood&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="kc">None</span><span class="p">,</span> <span class="mf">1.1</span> <span class="o">*</span> <span class="nb">max</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">dag_l</span><span class="p">))])</span>
        <span class="k">except</span> <span class="ne">FloatingPointError</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">dag_l</span><span class="p">,</span> <span class="s2">&quot;ko&quot;</span><span class="p">,</span> <span class="n">clip_on</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;gctree log-likelihood&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="kc">None</span><span class="p">,</span> <span class="mf">1.1</span> <span class="o">*</span> <span class="nb">max</span><span class="p">(</span><span class="n">dag_l</span><span class="p">)])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;parsimony tree&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dag_l</span><span class="p">)])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s2">&quot;out&quot;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s2">&quot;both&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span>
            <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s2">&quot;both&quot;</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="s2">&quot;off&quot;</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="s2">&quot;off&quot;</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="s2">&quot;off&quot;</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">outbase</span> <span class="o">+</span> <span class="s2">&quot;.inference.likelihood_rank.&quot;</span> <span class="o">+</span> <span class="n">img_type</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.n_topologies"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.n_topologies">[docs]</a>    <span class="nd">@_requires_dag</span>
    <span class="k">def</span> <span class="nf">n_topologies</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Count the number of topology classes, ignoring internal node</span>
<span class="sd">        sequences.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">count_topologies</span><span class="p">(</span><span class="n">collapse_leaves</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.iter_topology_classes"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.iter_topology_classes">[docs]</a>    <span class="nd">@_requires_dag</span>
    <span class="k">def</span> <span class="nf">iter_topology_classes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Sort trees by topology class.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A generator of CollapsedForest objects, each containing trees with the same topology,</span>
<span class="sd">                ignoring internal node labels. CollapsedForests will be yielded in reverse-order</span>
<span class="sd">                of the number of trees in each topology class, so that each CollapsedForest will</span>
<span class="sd">                contain at least as many trees as the one that follows.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">topology_dagfuncs</span> <span class="o">=</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">make_newickcountfuncs</span><span class="p">(</span>
            <span class="n">internal_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">collapse_leaves</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">ctopologies</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">weight_count</span><span class="p">(</span><span class="o">**</span><span class="n">topology_dagfuncs</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="c1"># yield classes with most members first</span>
        <span class="n">ctopologies</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="o">-</span><span class="n">t</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">ctopology</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">ctopologies</span><span class="p">:</span>
            <span class="n">tdag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tdag</span><span class="o">.</span><span class="n">trim_topology</span><span class="p">(</span><span class="n">ctopology</span><span class="p">,</span> <span class="n">collapse_leaves</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">yield</span> <span class="bp">self</span><span class="o">.</span><span class="n">_trimmed_self</span><span class="p">(</span><span class="n">tdag</span><span class="p">)</span></div>

<div class="viewcode-block" id="CollapsedForest.add_isotypes"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.add_isotypes">[docs]</a>    <span class="nd">@_requires_dag</span>
    <span class="k">def</span> <span class="nf">add_isotypes</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">isotypemap</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">isotypemap_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">idmap</span><span class="p">:</span> <span class="n">Mapping</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">idmap_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">isotype_names</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Adds isotype annotations, including inferred ancestral isotypes, to</span>
<span class="sd">        all nodes in stored trees.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">is_isotyped</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="n">iso_funcs</span> <span class="o">=</span> <span class="n">_isotype_annotation_dagfuncs</span><span class="p">(</span>
            <span class="n">isotypemap_file</span><span class="o">=</span><span class="n">isotypemap_file</span><span class="p">,</span>
            <span class="n">isotypemap</span><span class="o">=</span><span class="n">isotypemap</span><span class="p">,</span>
            <span class="n">idmap_file</span><span class="o">=</span><span class="n">idmap_file</span><span class="p">,</span>
            <span class="n">idmap</span><span class="o">=</span><span class="n">idmap</span><span class="p">,</span>
            <span class="n">isotype_names</span><span class="o">=</span><span class="n">isotype_names</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">def</span> <span class="nf">optimal_func</span><span class="p">(</span><span class="n">weightlist</span><span class="p">):</span>
            <span class="c1"># all weights (frozendicts containing isotypes) are equal</span>
            <span class="k">return</span> <span class="n">weightlist</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">optimal_weight_annotate</span><span class="p">(</span><span class="o">**</span><span class="n">iso_funcs</span><span class="p">,</span> <span class="n">optimal_func</span><span class="o">=</span><span class="n">optimal_func</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">preorder</span><span class="p">(</span><span class="n">skip_root</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;isotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">_dp_data</span></div>

<div class="viewcode-block" id="CollapsedForest.sample_tree"><a class="viewcode-back" href="../../stubs/gctree.branching_processes.CollapsedForest.html#gctree.CollapsedForest.sample_tree">[docs]</a>    <span class="k">def</span> <span class="nf">sample_tree</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">CollapsedTree</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Sample a random CollapsedTree from the forest.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_clade_tree_to_ctree</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">sample</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot sample trees from an empty forest&quot;</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="nf">_trimmed_self</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dag</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create a new CollapsedForest object from a subset of the current</span>
<span class="sd">        one&#39;s history DAG, properly initializing validation_stats and</span>
<span class="sd">        n_trees.&quot;&quot;&quot;</span>
        <span class="n">newforest</span> <span class="o">=</span> <span class="n">CollapsedForest</span><span class="p">()</span>
        <span class="n">newforest</span><span class="o">.</span><span class="n">_validation_stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">newforest</span><span class="o">.</span><span class="n">n_trees</span> <span class="o">=</span> <span class="n">dag</span><span class="o">.</span><span class="n">count_trees</span><span class="p">()</span>
        <span class="n">newforest</span><span class="o">.</span><span class="n">_forest</span> <span class="o">=</span> <span class="n">dag</span>
        <span class="n">newforest</span><span class="o">.</span><span class="n">parameters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span>
        <span class="k">return</span> <span class="n">newforest</span>

    <span class="k">def</span> <span class="nf">_clade_tree_to_ctree</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">clade_tree</span><span class="p">:</span> <span class="n">hdag</span><span class="o">.</span><span class="n">HistoryDag</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">CollapsedTree</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Create and validate :meth:`CollapsedTree` object from tree-shaped</span>
<span class="sd">        history DAG.</span>

<span class="sd">        Uses self._validation_stats dictionary, if available, for validation. Dictionary keys are:</span>
<span class="sd">            root: The expected root name for the resulting tree (for validation)</span>
<span class="sd">            counts: Expected node abundances for each node name (for validation)</span>
<span class="sd">            parsimony_score: Expected Hamming parsimony score (for validation)</span>
<span class="sd">            root_seq: Expected root sequence (for validation)</span>
<span class="sd">            leaf_seqs: Expected leaf names, keyed by leaf sequences (for validation)</span>

<span class="sd">        Args:</span>
<span class="sd">            clade_tree: A tree-shaped history DAG, like that returned by :meth:`historydag.HistoryDag.sample`</span>

<span class="sd">        Returns:</span>
<span class="sd">            :meth:`CollapsedTree` object matching the topology of ``clade_tree``, but fully collapsed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">etetree</span> <span class="o">=</span> <span class="n">clade_tree</span><span class="o">.</span><span class="n">to_ete</span><span class="p">(</span>
            <span class="n">name_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">n</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">],</span>
            <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span>
            <span class="n">feature_funcs</span><span class="o">=</span><span class="p">{</span>
                <span class="s2">&quot;abundance&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">n</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">],</span>
                <span class="s2">&quot;isotype&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">n</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;isotype&quot;</span><span class="p">],</span>
            <span class="p">},</span>
        <span class="p">)</span>

        <span class="c1"># Remove dummy leaf added below root for hDAG compatibility</span>
        <span class="n">dummyleaves</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">node</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">etetree</span><span class="o">.</span><span class="n">children</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">()</span> <span class="ow">and</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dummyleaves</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;Multiple temporary leaves found in tree. Does an observed sequence have name &#39;&#39;?&quot;</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">dummyleaves</span><span class="p">:</span>
            <span class="n">leaf</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">prevent_nondicotomic</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">ctree</span> <span class="o">=</span> <span class="n">CollapsedTree</span><span class="p">(</span><span class="n">etetree</span><span class="p">)</span>

        <span class="c1"># Fix internal node names to be unique, and verify</span>
        <span class="c1"># The maps from nodes to names and nodes to sequences are bijections</span>
        <span class="n">n_nodes</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">seqs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="n">n_nodes</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">names</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
            <span class="n">seqs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">n_names</span><span class="p">,</span> <span class="n">n_seqs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">names</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">n_nodes</span> <span class="o">==</span> <span class="n">n_names</span> <span class="ow">and</span> <span class="n">n_names</span> <span class="o">==</span> <span class="n">n_seqs</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;Multiple sequences with the same name, or multiple&quot;</span>
                <span class="s2">&quot;names for the same sequence, observed in collapsed tree.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Here can do some validation on the tree:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># root name:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s2">&quot;root&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;collapsed tree should have root name &#39;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s1">&#39;root&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&#39; but has instead </span><span class="si">{</span><span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
            <span class="c1"># counts:</span>
            <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">etetree</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="n">counts</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">==</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="n">counts</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">==</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">assert</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">==</span> <span class="mi">0</span>

            <span class="c1"># unnamed_seq issue:</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;unnamed_seq&quot;</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Some node names are missing&quot;</span><span class="p">)</span>

            <span class="c1"># Parsimony:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s2">&quot;parsimony_score&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="nb">sum</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">gctree</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">hamming_distance</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">up</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">()</span>
                <span class="p">]</span>
            <span class="p">):</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;History DAG tree parsimony score does not match parsimony score provided&quot;</span>
                <span class="p">)</span>
            <span class="c1"># Root sequence:</span>
            <span class="k">if</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">sequence</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s2">&quot;root_seq&quot;</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;History DAG root node sequence does not match root sequence provided&quot;</span>
                <span class="p">)</span>
            <span class="c1"># Leaf names:</span>
            <span class="n">leaf_seqs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_stats</span><span class="p">[</span><span class="s2">&quot;leaf_seqs&quot;</span><span class="p">]</span>
            <span class="c1"># A dictionary of leaf sequences to leaf names</span>
            <span class="n">observed_set</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">node</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">()</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="p">}</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">observed_set</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_root</span><span class="p">()</span> <span class="ow">and</span> <span class="n">leaf_seqs</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span> <span class="o">!=</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                        <span class="s2">&quot;History DAG tree leaf names don&#39;t match sequences&quot;</span>
                    <span class="p">)</span>
            <span class="n">observed_seqs</span> <span class="o">=</span> <span class="p">{</span><span class="n">node</span><span class="o">.</span><span class="n">sequence</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">observed_set</span><span class="p">}</span>
            <span class="n">nonroot_observed_seqs</span> <span class="o">=</span> <span class="n">observed_seqs</span> <span class="o">-</span> <span class="p">{</span><span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">sequence</span><span class="p">}</span>
            <span class="n">nonroot_leaf_seqs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">leaf_seqs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">-</span> <span class="p">{</span><span class="n">ctree</span><span class="o">.</span><span class="n">tree</span><span class="o">.</span><span class="n">sequence</span><span class="p">}</span>
            <span class="k">if</span> <span class="n">nonroot_leaf_seqs</span> <span class="o">!=</span> <span class="n">nonroot_observed_seqs</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;Observed nonroot sequences in history DAG tree don&#39;t match &quot;</span>
                    <span class="s2">&quot;observed nonroot sequences passed in leaf_seqs.&quot;</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;No validation was performed on tree&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">ctree</span>

    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;return a string representation for printing.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;n_trees = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">n_trees</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">])</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">yield from</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ctrees</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">cladetree</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_forest</span><span class="o">.</span><span class="n">get_trees</span><span class="p">():</span>
                <span class="k">yield</span> <span class="bp">self</span><span class="o">.</span><span class="n">_clade_tree_to_ctree</span><span class="p">(</span><span class="n">cladetree</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield from</span> <span class="p">()</span>

    <span class="k">def</span> <span class="nf">__getstate__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Avoid pickling large cached abundance data.</span>
        <span class="c1"># hDAG also defines its own getstate.</span>
        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">d</span><span class="p">[</span><span class="s2">&quot;_cm_countlist&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">d</span></div>


<span class="k">def</span> <span class="nf">_mle_helper</span><span class="p">(</span>
    <span class="n">ll</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]],</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
    <span class="c1"># initialization</span>
    <span class="n">x_0</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>
    <span class="n">bounds</span> <span class="o">=</span> <span class="p">((</span><span class="mf">1e-6</span><span class="p">,</span> <span class="mi">1</span> <span class="o">-</span> <span class="mf">1e-6</span><span class="p">),</span> <span class="p">(</span><span class="mf">1e-6</span><span class="p">,</span> <span class="mi">1</span> <span class="o">-</span> <span class="mf">1e-6</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">f</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;negative log likelihood.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="o">-</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">ll</span><span class="p">(</span><span class="o">*</span><span class="n">x</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

    <span class="n">grad_check</span> <span class="o">=</span> <span class="n">sco</span><span class="o">.</span><span class="n">check_grad</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="n">x_0</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">grad_check</span> <span class="o">&gt;</span> <span class="mf">1e-3</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;gradient mismatches finite difference approximation by </span><span class="si">{</span><span class="n">grad_check</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="ne">RuntimeWarning</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">sco</span><span class="o">.</span><span class="n">minimize</span><span class="p">(</span>
        <span class="n">f</span><span class="p">,</span>
        <span class="n">jac</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">x0</span><span class="o">=</span><span class="n">x_0</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s2">&quot;L-BFGS-B&quot;</span><span class="p">,</span>
        <span class="n">options</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ftol&quot;</span><span class="p">:</span> <span class="mf">1e-10</span><span class="p">},</span>
        <span class="n">bounds</span><span class="o">=</span><span class="n">bounds</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># update params if None and optimization successful</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="o">.</span><span class="n">success</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;optimization not sucessful, &quot;</span> <span class="o">+</span> <span class="n">result</span><span class="o">.</span><span class="n">message</span><span class="p">,</span> <span class="ne">RuntimeWarning</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">result</span><span class="o">.</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">_lltree</span><span class="p">(</span><span class="n">cm_counts</span><span class="p">,</span> <span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Log likelihood of branching process parameters :math:`(p, q)`</span>
<span class="sd">    .. math::</span>
<span class="sd">        \ell(p, q; T, A) = \log\mathbb{P}(T, A \mid p, q)</span>

<span class="sd">    Args:</span>
<span class="sd">        cm_counts: an iterable containing tuples `((c, m), n)` where `n` is the number of nodes</span>
<span class="sd">            in the tree with abundance `c` and `m` mutant clades</span>
<span class="sd">        p: branching probability</span>
<span class="sd">        q: mutation probability</span>
<span class="sd">    Returns:</span>
<span class="sd">        Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">count_ls</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">cm</span><span class="p">,</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">cm_counts</span><span class="p">]</span>
    <span class="n">cm_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">cm</span> <span class="k">for</span> <span class="n">cm</span><span class="p">,</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">cm_counts</span><span class="p">]</span>
    <span class="c1"># extract vector of function values and gradient components</span>
    <span class="n">logf_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">cm_list</span><span class="p">]</span>
    <span class="n">logf</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">count_ls</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">logf_data</span><span class="p">)])</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">grad_ll_genotype</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">count_ls</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">logf_data</span><span class="p">)]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">logf</span><span class="p">,</span> <span class="n">grad_ll_genotype</span>


<span class="k">def</span> <span class="nf">_make_dag</span><span class="p">(</span><span class="n">trees</span><span class="p">,</span> <span class="n">from_copy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Build a history DAG from ambiguous or disambiguated trees, whose nodes</span>
<span class="sd">    have abundance, name, and sequence attributes.&quot;&quot;&quot;</span>
    <span class="c1"># preprocess trees so they&#39;re acceptable inputs</span>
    <span class="c1"># Assume all trees have fixed root sequence and fixed leaf sequences</span>
    <span class="k">def</span> <span class="nf">get_sequence</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">base</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gctree</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">bases</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unrecognized base found in node &#39;</span><span class="si">{</span><span class="n">node</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&#39;. &quot;</span>
                <span class="s2">&quot;Ambiguous bases are not permitted in observed sequences.&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">node</span><span class="o">.</span><span class="n">sequence</span>

    <span class="n">leaf_seqs</span> <span class="o">=</span> <span class="p">{</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">node</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()}</span>

    <span class="c1"># Assume all trees have same observed nodes.</span>
    <span class="n">sequence_counts</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">node</span><span class="o">.</span><span class="n">sequence</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">traverse</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">abundance</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="n">from_copy</span><span class="p">:</span>
        <span class="n">trees</span> <span class="o">=</span> <span class="p">[</span><span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">children</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">):</span>
        <span class="k">pass</span>  <span class="c1"># We&#39;re all good!</span>
    <span class="k">elif</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leaf_seqs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sequence_counts</span><span class="p">:</span>
            <span class="n">sequence_counts</span><span class="p">[</span><span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">:</span>
            <span class="n">newleaf</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">newleaf</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="n">trees</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sequence</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">tree</span><span class="o">.</span><span class="n">sequence</span> <span class="o">!=</span> <span class="n">newleaf</span><span class="o">.</span><span class="n">sequence</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;At least some trees unifurcate at root, but root sequence is not fixed.&quot;</span>
                <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># This should never happen in parsimony setting, when internal edges</span>
        <span class="c1"># are collapsed by sequence</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="s2">&quot;Root sequence observed, but the corresponding leaf is not a child of the root node. &quot;</span>
            <span class="s2">&quot;Gctree inference may give nonsensical results. Are you sure these are parsimony trees?&quot;</span>
        <span class="p">)</span>

    <span class="n">dag</span> <span class="o">=</span> <span class="n">hdag</span><span class="o">.</span><span class="n">history_dag_from_etes</span><span class="p">(</span>
        <span class="n">trees</span><span class="p">,</span>
        <span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span>
        <span class="n">attr_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;isotype&quot;</span><span class="p">:</span> <span class="n">frozendict</span><span class="p">(),</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="c1"># If there are too many ambiguities at too many nodes, disambiguation will</span>
    <span class="c1"># hang. Need to have an alternative (disambiguate each tree before putting in dag):</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">dag</span><span class="o">.</span><span class="n">count_trees</span><span class="p">(</span><span class="n">expand_count_func</span><span class="o">=</span><span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">sequence_resolutions_count</span><span class="p">)</span>
        <span class="o">/</span> <span class="n">dag</span><span class="o">.</span><span class="n">count_trees</span><span class="p">()</span>
        <span class="o">&gt;</span> <span class="mi">500000000</span>
    <span class="p">):</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Parsimony trees have too many ambiguities for disambiguation in history DAG. &quot;</span>
            <span class="s2">&quot;Disambiguating trees individually. History DAG may find fewer new parsimony trees.&quot;</span>
        <span class="p">)</span>
        <span class="n">distrees</span> <span class="o">=</span> <span class="p">[</span><span class="n">disambiguate</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="k">for</span> <span class="n">tree</span> <span class="ow">in</span> <span class="n">trees</span><span class="p">]</span>
        <span class="n">dag</span> <span class="o">=</span> <span class="n">hdag</span><span class="o">.</span><span class="n">history_dag_from_etes</span><span class="p">(</span>
            <span class="n">distrees</span><span class="p">,</span>
            <span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">],</span>
            <span class="n">attr_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="s2">&quot;isotype&quot;</span><span class="p">:</span> <span class="n">frozendict</span><span class="p">(),</span>
            <span class="p">},</span>
        <span class="p">)</span>
    <span class="n">dag</span><span class="o">.</span><span class="n">explode_nodes</span><span class="p">(</span><span class="n">expand_func</span><span class="o">=</span><span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">sequence_resolutions</span><span class="p">)</span>
    <span class="c1"># Look for (even) more trees:</span>
    <span class="n">dag</span><span class="o">.</span><span class="n">add_all_allowed_edges</span><span class="p">(</span><span class="n">adjacent_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">dag</span><span class="o">.</span><span class="n">trim_optimal_weight</span><span class="p">()</span>
    <span class="n">dag</span><span class="o">.</span><span class="n">convert_to_collapsed</span><span class="p">()</span>
    <span class="c1"># Add abundances to attrs:</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">dag</span><span class="o">.</span><span class="n">preorder</span><span class="p">(</span><span class="n">skip_root</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">in</span> <span class="n">sequence_counts</span><span class="p">:</span>
            <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequence_counts</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;sequence_counts dictionary should contain all leaf sequences&quot;</span>
                <span class="p">)</span>
            <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dag</span><span class="o">.</span><span class="n">hamming_parsimony_count</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;History DAG parsimony search resulted in parsimony trees of unexpected weights:</span><span class="se">\n</span><span class="s2"> </span><span class="si">{</span><span class="n">dag</span><span class="o">.</span><span class="n">hamming_parsimony_count</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">dag</span><span class="o">.</span><span class="n">preorder</span><span class="p">(</span><span class="n">skip_root</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">node</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">()</span>
            <span class="ow">and</span> <span class="nb">frozenset</span><span class="p">({</span><span class="n">node</span><span class="o">.</span><span class="n">label</span><span class="p">})</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">clades</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;An internal node not adjacent to a leaf with the same label was found with nonzero abundance.&quot;</span>
            <span class="p">)</span>

    <span class="c1"># names on internal nodes are all messed up from disambiguation step, we&#39;ll</span>
    <span class="c1"># fix them in CollapsedTree.__init__.</span>
    <span class="k">return</span> <span class="n">dag</span>


<span class="k">def</span> <span class="nf">_cmcounter_dagfuncs</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Functions for accumulating frozen multisets of (c, m) pairs in trees in</span>
<span class="sd">    the DAG.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">edge_weight_func</span><span class="p">(</span><span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">n1</span><span class="o">.</span><span class="n">label</span> <span class="o">==</span> <span class="n">n2</span><span class="o">.</span><span class="n">label</span> <span class="ow">and</span> <span class="n">n2</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
            <span class="c1"># Then this is a leaf-adjacent node with nonzero abundance</span>
            <span class="k">return</span> <span class="n">multiset</span><span class="o">.</span><span class="n">FrozenMultiset</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">n2</span><span class="o">.</span><span class="n">clades</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">frozenset</span><span class="p">({</span><span class="n">n2</span><span class="o">.</span><span class="n">label</span><span class="p">})</span> <span class="ow">in</span> <span class="n">n2</span><span class="o">.</span><span class="n">clades</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">-=</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">multiset</span><span class="o">.</span><span class="n">FrozenMultiset</span><span class="p">([(</span><span class="n">n2</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">],</span> <span class="n">m</span><span class="p">)])</span>

    <span class="k">def</span> <span class="nf">accum_func</span><span class="p">(</span><span class="n">cmsetlist</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">multiset</span><span class="o">.</span><span class="n">FrozenMultiset</span><span class="p">]):</span>
        <span class="n">st</span> <span class="o">=</span> <span class="n">multiset</span><span class="o">.</span><span class="n">FrozenMultiset</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">cmset</span> <span class="ow">in</span> <span class="n">cmsetlist</span><span class="p">:</span>
            <span class="n">st</span> <span class="o">+=</span> <span class="n">cmset</span>
        <span class="k">return</span> <span class="n">st</span>

    <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;start_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">multiset</span><span class="o">.</span><span class="n">FrozenMultiset</span><span class="p">(),</span>
            <span class="s2">&quot;edge_weight_func&quot;</span><span class="p">:</span> <span class="n">edge_weight_func</span><span class="p">,</span>
            <span class="s2">&quot;accum_func&quot;</span><span class="p">:</span> <span class="n">accum_func</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;cmcounters&quot;</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">_ll_genotype_dagfuncs</span><span class="p">(</span><span class="n">p</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">q</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Return functions for counting tree log likelihood on the history DAG.</span>

<span class="sd">    For numerical consistency, we resort to the use of ``decimal.Decimal``.</span>
<span class="sd">    This is exactly for the purpose of solving the problem that float sum is</span>
<span class="sd">    sensitive to order of summation, while Decimal sum is not.</span>

<span class="sd">    Still, there seems to be some inconsistency (probably in the part of the computation done with floats)</span>
<span class="sd">    so we wrap the Decimal in a :class:`historydag.utils.FloatState` object, where the exposed</span>
<span class="sd">    float is a rounded version of the hidden Decimal state, which is used for actual computations.</span>


<span class="sd">    Args:</span>
<span class="sd">        p, q: branching process parameters</span>

<span class="sd">    Returns:</span>
<span class="sd">        A :meth:`historydag.utils.AddFuncDict` which may be passed as keyword arguments</span>
<span class="sd">        to :meth:`historydag.HistoryDag.weight_count`, :meth:`historydag.HistoryDag.trim_optimal_weight`,</span>
<span class="sd">        or :meth:`historydag.HistoryDag.optimal_weight_annotate`</span>
<span class="sd">        methods to trim or annotate a :meth:`historydag.HistoryDag` according to branching process likelihood.</span>
<span class="sd">        Weight format is ``decimal.Decimal``.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">edge_weight_ll_genotype</span><span class="p">(</span><span class="n">n1</span><span class="p">:</span> <span class="n">hdag</span><span class="o">.</span><span class="n">HistoryDagNode</span><span class="p">,</span> <span class="n">n2</span><span class="p">:</span> <span class="n">hdag</span><span class="o">.</span><span class="n">HistoryDagNode</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The _ll_genotype weight of the target node, unless it should be</span>
<span class="sd">        collapsed, then 0.</span>

<span class="sd">        Expects DAG to have abundances added so that each node has</span>
<span class="sd">        &quot;abundance&quot; key in attr dict.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">n2</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">()</span> <span class="ow">and</span> <span class="n">n2</span><span class="o">.</span><span class="n">label</span> <span class="o">==</span> <span class="n">n1</span><span class="o">.</span><span class="n">label</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">FloatState</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">state</span><span class="o">=</span><span class="n">Decimal</span><span class="p">(</span><span class="mf">0.0</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">n2</span><span class="o">.</span><span class="n">clades</span><span class="p">)</span>
            <span class="c1"># Check if this edge should be collapsed, and reduce mutant descendants</span>
            <span class="k">if</span> <span class="nb">frozenset</span><span class="p">({</span><span class="n">n2</span><span class="o">.</span><span class="n">label</span><span class="p">})</span> <span class="ow">in</span> <span class="n">n2</span><span class="o">.</span><span class="n">clades</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">-=</span> <span class="mi">1</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">n2</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;abundance&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">n1</span><span class="o">.</span><span class="n">is_root</span><span class="p">()</span> <span class="ow">and</span> <span class="n">c</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Add pseudocount for unobserved root unifurcation</span>
                <span class="n">c</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">Decimal</span><span class="p">(</span><span class="n">CollapsedTree</span><span class="o">.</span><span class="n">_ll_genotype</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">FloatState</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="mi">8</span><span class="p">)),</span> <span class="n">state</span><span class="o">=</span><span class="n">res</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">accum_func</span><span class="p">(</span><span class="n">weightlist</span><span class="p">):</span>
        <span class="n">res</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">weight</span><span class="o">.</span><span class="n">state</span> <span class="k">for</span> <span class="n">weight</span> <span class="ow">in</span> <span class="n">weightlist</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">FloatState</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="mi">8</span><span class="p">)),</span> <span class="n">state</span><span class="o">=</span><span class="n">res</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;start_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">FloatState</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">state</span><span class="o">=</span><span class="n">Decimal</span><span class="p">(</span><span class="mi">0</span><span class="p">)),</span>
            <span class="s2">&quot;edge_weight_func&quot;</span><span class="p">:</span> <span class="n">edge_weight_ll_genotype</span><span class="p">,</span>
            <span class="s2">&quot;accum_func&quot;</span><span class="p">:</span> <span class="n">accum_func</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Log Likelihood&quot;</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">_allele_dagfuncs</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Return functions for filtering trees in a history DAG by allele count.</span>

<span class="sd">    The number of alleles in a tree is the number of unique sequences observed on nodes of that tree.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A :meth:`historydag.utils.AddFuncDict` which may be passed as keyword arguments</span>
<span class="sd">        to :meth:`historydag.HistoryDag.weight_count`, :meth:`historydag.HistoryDag.trim_optimal_weight`,</span>
<span class="sd">        or :meth:`historydag.HistoryDag.optimal_weight_annotate`</span>
<span class="sd">        methods to trim or annotate a :meth:`historydag.HistoryDag` according to allele count.</span>
<span class="sd">        Weight format is ``int``.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;start_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;edge_weight_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n1</span><span class="p">,</span> <span class="n">n2</span><span class="p">:</span> <span class="n">n1</span><span class="o">.</span><span class="n">label</span> <span class="o">!=</span> <span class="n">n2</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
            <span class="s2">&quot;accum_func&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Alleles&quot;</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, William DeWitt.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>