<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gctree.mutation_model &mdash; gctree  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html">
            <img src="../../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cite.html">Citing gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../rendering-demo.html">Custom tree visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">CLI Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">gctree</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html#parsimony-utilities">Parsimony utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html#isotype">isotype</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html#modules">Modules</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html#developer-tools">Developer tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../developer.html#todo-list">Todo list</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gctree</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>gctree.mutation_model</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gctree.mutation_model</h1><div class="highlight"><pre>
<span></span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Mutation models.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">gctree.utils</span> <span class="k">as</span> <span class="nn">utils</span>

<span class="kn">from</span> <span class="nn">ete3</span> <span class="kn">import</span> <span class="n">TreeNode</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">poisson</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">import</span> <span class="nn">historydag</span> <span class="k">as</span> <span class="nn">hdag</span>
<span class="kn">from</span> <span class="nn">multiset</span> <span class="kn">import</span> <span class="n">FrozenMultiset</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Callable</span>


<div class="viewcode-block" id="MutationModel"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.MutationModel.html#gctree.mutation_model.MutationModel">[docs]</a><span class="k">class</span> <span class="nc">MutationModel</span><span class="p">:</span>
    <span class="sa">r</span><span class="sd">&quot;&quot;&quot;A class for a mutation model, and functions to mutate sequences.</span>

<span class="sd">    Args:</span>
<span class="sd">        mutability_file: S5F format mutabilities</span>
<span class="sd">        substitution_file: S5F format substitution biases</span>
<span class="sd">        mutation_order: whether or not to mutate sequences using a context sensitive manner</span>
<span class="sd">                        where mutation order matters</span>
<span class="sd">        with_replacement: allow the same position to mutate multiple times on a single branch</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="MutationModel.__init__"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.html#gctree.mutation_model.MutationModel.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">mutability_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">substitution_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">mutation_order</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">with_replacement</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mutation_order</span> <span class="o">=</span> <span class="n">mutation_order</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">with_replacement</span> <span class="o">=</span> <span class="n">with_replacement</span>
        <span class="k">if</span> <span class="n">mutability_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">substitution_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mutability_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="c1"># eat header</span>
                <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">motif</span><span class="p">,</span> <span class="n">score</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="mi">2</span><span class="p">]</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

            <span class="c1"># kmer k</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">substitution_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="c1"># eat header</span>
                <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                    <span class="n">motif</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span>
                        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span><span class="p">[</span><span class="n">motif</span><span class="p">],</span>
                        <span class="p">{</span><span class="n">b</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span><span class="p">])},</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span> <span class="o">=</span> <span class="kc">None</span></div>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_disambiguate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;generator of all possible nt sequences implied by a sequence</span>
<span class="sd">        containing ambiguous bases.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">_sequence_disambiguations</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

<div class="viewcode-block" id="MutationModel.mutability"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.MutationModel.html#gctree.mutation_model.MutationModel.mutability">[docs]</a>    <span class="k">def</span> <span class="nf">mutability</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kmer</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Returns the mutability of a central base of :math:`k`-mer, along with</span>
<span class="sd">        nucleotide bias averages over ambiguous ``&quot;N&quot;`` nucleotide identities.</span>

<span class="sd">        Args:</span>
<span class="sd">            kmer: nucleotide :math:`k`-mer</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;kmer mutability only defined for context models&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kmer</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;kmer of length </span><span class="si">{}</span><span class="s2"> inconsistent with context model kmer length </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="nb">len</span><span class="p">(</span><span class="n">kmer</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">n</span> <span class="ow">in</span> <span class="s2">&quot;ACGTN&quot;</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">kmer</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;sequence </span><span class="si">{}</span><span class="s2"> must contain only characters A, C, G, T, or N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">kmer</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="n">mutabilities_to_average</span><span class="p">,</span> <span class="n">substitutions_to_average</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">context_model</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">MutationModel</span><span class="o">.</span><span class="n">_disambiguate</span><span class="p">(</span><span class="n">kmer</span><span class="p">)]</span>
        <span class="p">)</span>

        <span class="n">average_mutability</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mutabilities_to_average</span><span class="p">)</span>
        <span class="n">average_substitution</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">b</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span>
                <span class="n">substitution_dict</span><span class="p">[</span><span class="n">b</span><span class="p">]</span> <span class="k">for</span> <span class="n">substitution_dict</span> <span class="ow">in</span> <span class="n">substitutions_to_average</span>
            <span class="p">)</span>
            <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">substitutions_to_average</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="s2">&quot;ACGT&quot;</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">average_mutability</span><span class="p">,</span> <span class="n">average_substitution</span></div>

<div class="viewcode-block" id="MutationModel.mutabilities"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.MutationModel.html#gctree.mutation_model.MutationModel.mutabilities">[docs]</a>    <span class="k">def</span> <span class="nf">mutabilities</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]]:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Returns the mutability of a sequence at each site, along with</span>
<span class="sd">        nucleotide biases.</span>

<span class="sd">        Args:</span>
<span class="sd">            sequence: nucleotide sequence</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">context_model</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">dict</span><span class="p">((</span><span class="n">n2</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="k">if</span> <span class="n">n2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">n</span> <span class="k">else</span> <span class="p">(</span><span class="n">n2</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span> <span class="k">for</span> <span class="n">n2</span> <span class="ow">in</span> <span class="s2">&quot;ACGT&quot;</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">sequence</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># pad with Ns to allow averaged edge effects</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">sequence</span> <span class="o">+</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
            <span class="c1"># mutabilities of each nucleotide</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">mutability</span><span class="p">(</span><span class="n">sequence</span><span class="p">[(</span><span class="n">i</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="p">:</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)])</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">k</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
            <span class="p">]</span></div>

<div class="viewcode-block" id="MutationModel.mutate"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.MutationModel.html#gctree.mutation_model.MutationModel.mutate">[docs]</a>    <span class="k">def</span> <span class="nf">mutate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">lambda0</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">frame</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Mutate a sequence, with lamdba0 the baseline mutability. Cannot</span>
<span class="sd">        mutate the same position multiple times.</span>

<span class="sd">        Args:</span>
<span class="sd">            sequence: nucleotide sequence to mutate</span>
<span class="sd">            lambda0: a baseline mutation rate</span>
<span class="sd">            frame: the reading frame of the first postition</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequence_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">frame</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">codon_start</span> <span class="o">=</span> <span class="n">frame</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">codon_end</span> <span class="o">=</span> <span class="n">codon_start</span> <span class="o">+</span> <span class="mi">3</span> <span class="o">*</span> <span class="p">((</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">codon_start</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">codon_start</span><span class="p">:</span><span class="n">codon_end</span><span class="p">])</span><span class="o">.</span><span class="n">translate</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;sequence contains stop codon!&quot;</span><span class="p">)</span>

        <span class="n">mutabilities</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutabilities</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">sequence_mutability</span> <span class="o">=</span> <span class="p">(</span>
            <span class="nb">sum</span><span class="p">(</span><span class="n">mutability</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">mutability</span> <span class="ow">in</span> <span class="n">mutabilities</span><span class="p">)</span> <span class="o">/</span> <span class="n">sequence_length</span>
        <span class="p">)</span>
        <span class="c1"># poisson rate for this sequence (given its relative mutability)</span>
        <span class="n">lambda_sequence</span> <span class="o">=</span> <span class="n">sequence_mutability</span> <span class="o">*</span> <span class="n">lambda0</span>
        <span class="c1"># number of mutations m</span>
        <span class="n">trials</span> <span class="o">=</span> <span class="mi">20</span>
        <span class="k">for</span> <span class="n">trial</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">trials</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lambda_sequence</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">m</span> <span class="o">&lt;=</span> <span class="n">sequence_length</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">with_replacement</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="n">trial</span> <span class="o">==</span> <span class="n">trials</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;mutations saturating, consider reducing lambda0&quot;</span><span class="p">)</span>

        <span class="c1"># mutate the sites with mutations</span>
        <span class="c1"># if frame is not None and sequence contains stop codon, try again, up to 10 times</span>
        <span class="n">unmutated_positions</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">):</span>
            <span class="n">sequence_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>  <span class="c1"># make string a list so we can modify it</span>
            <span class="c1"># Determine the position to mutate from the mutability matrix</span>
            <span class="n">mutability_p</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span><span class="n">mutabilities</span><span class="p">[</span><span class="n">pos</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">unmutated_positions</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">trial</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">trials</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">mut_pos</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                    <span class="n">unmutated_positions</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">mutability_p</span> <span class="o">/</span> <span class="n">mutability_p</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="p">)</span>
                <span class="c1"># Now draw the target nucleotide using the substitution matrix</span>
                <span class="n">substitution_p</span> <span class="o">=</span> <span class="p">[</span><span class="n">mutabilities</span><span class="p">[</span><span class="n">mut_pos</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">n</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="s2">&quot;ACGT&quot;</span><span class="p">]</span>
                <span class="k">assert</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">substitution_p</span><span class="p">)</span> <span class="o">-</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mf">1e-10</span>
                <span class="n">chosen_target</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">substitution_p</span><span class="p">)</span>
                <span class="n">original_base</span> <span class="o">=</span> <span class="n">sequence_list</span><span class="p">[</span><span class="n">mut_pos</span><span class="p">]</span>
                <span class="n">sequence_list</span><span class="p">[</span><span class="n">mut_pos</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ACGT&quot;</span><span class="p">[</span><span class="n">chosen_target</span><span class="p">]</span>
                <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence_list</span><span class="p">)</span>  <span class="c1"># reconstruct our sequence</span>
                <span class="k">if</span> <span class="p">(</span>
                    <span class="n">frame</span> <span class="ow">is</span> <span class="kc">None</span>
                    <span class="ow">or</span> <span class="s2">&quot;*&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">codon_start</span><span class="p">:</span><span class="n">codon_end</span><span class="p">])</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
                <span class="p">):</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutation_order</span><span class="p">:</span>
                        <span class="c1"># if mutation order matters, the mutabilities of the sequence need to be updated</span>
                        <span class="n">mutabilities</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutabilities</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">with_replacement</span><span class="p">:</span>
                        <span class="c1"># Remove this position so we don&#39;t mutate it again</span>
                        <span class="n">unmutated_positions</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">mut_pos</span><span class="p">)</span>
                    <span class="k">break</span>
                <span class="k">if</span> <span class="n">trial</span> <span class="o">==</span> <span class="n">trials</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                        <span class="s2">&quot;stop codon in simulated sequence on &quot;</span>
                        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">trials</span><span class="p">)</span>
                        <span class="o">+</span> <span class="s2">&quot; consecutive attempts&quot;</span>
                    <span class="p">)</span>
                <span class="n">sequence_list</span><span class="p">[</span>
                    <span class="n">mut_pos</span>
                <span class="p">]</span> <span class="o">=</span> <span class="n">original_base</span>  <span class="c1"># &lt;-- we only get here if we are retrying</span>

        <span class="k">return</span> <span class="n">sequence</span></div>

<div class="viewcode-block" id="MutationModel.simulate"><a class="viewcode-back" href="../../stubs/gctree.mutation_model.MutationModel.html#gctree.mutation_model.MutationModel.simulate">[docs]</a>    <span class="k">def</span> <span class="nf">simulate</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">sequence</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">seq_bounds</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">fitness_function</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">seq</span><span class="p">:</span> <span class="mf">0.9</span><span class="p">,</span>
        <span class="n">lambda0</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="n">frame</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">N_init</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">N</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">T</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">TreeNode</span><span class="p">:</span>
        <span class="sa">r</span><span class="sd">&quot;&quot;&quot;Simulate a neutral binary branching process with the mutation model, returning a :class:`ete3.Treenode` object.</span>

<span class="sd">        Args:</span>
<span class="sd">            sequence: root nucleotide sequence</span>
<span class="sd">            seq_bounds: ranges for two subsequences used as two parallel genes</span>
<span class="sd">            fitness_function: mean number offspring as a function of sequence</span>
<span class="sd">            lambda0: baseline mutation rate(s)</span>
<span class="sd">            frame: coding frame of starting position(s)</span>
<span class="sd">            N_init: initial naive abundnace</span>
<span class="sd">            N: maximum population size</span>
<span class="sd">            T: maximum generation time</span>
<span class="sd">            n: sample size</span>
<span class="sd">            verbose: print more messages</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Checking the validity of the input parameters:</span>
        <span class="k">if</span> <span class="n">N</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">T</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Only one of N and T can be used. One must be None.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">N</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">T</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Either N or T must be specified.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">N</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="n">N</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;n (</span><span class="si">{}</span><span class="s2">) must not larger than N (</span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>

        <span class="c1"># Planting the tree:</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="p">()</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;terminated&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;abundance&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;time&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="c1"># add fitness attribute, interpreted as mean of offspring distribution</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">,</span> <span class="n">fitness_function</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">N_init</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N_init</span><span class="p">):</span>
                <span class="n">child</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="p">()</span>
                <span class="n">child</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
                <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;abundance&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;terminated&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;time&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="c1"># add fitness attribute, interpreted as mean of offspring distribution</span>
                <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">,</span> <span class="n">fitness_function</span><span class="p">(</span><span class="n">child</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
                <span class="n">tree</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>

        <span class="n">t</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># &lt;-- time</span>
        <span class="n">leaves_unterminated</span> <span class="o">=</span> <span class="n">N_init</span>
        <span class="k">while</span> <span class="p">(</span>
            <span class="n">leaves_unterminated</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="ow">and</span> <span class="p">(</span><span class="n">leaves_unterminated</span> <span class="o">&lt;</span> <span class="n">N</span> <span class="k">if</span> <span class="n">N</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">True</span><span class="p">)</span>
            <span class="ow">and</span> <span class="p">(</span><span class="n">t</span> <span class="o">&lt;</span> <span class="nb">max</span><span class="p">(</span><span class="n">T</span><span class="p">)</span> <span class="k">if</span> <span class="n">T</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">True</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;At time:&quot;</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
            <span class="n">t</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">list_of_leaves</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">())</span>
            <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">list_of_leaves</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">list_of_leaves</span><span class="p">:</span>
                <span class="c1"># add fitness attribute, interpreted as mean of offspring distribution</span>
                <span class="n">leaf</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;fitness&quot;</span><span class="p">,</span> <span class="n">fitness_function</span><span class="p">(</span><span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span><span class="p">))</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">leaf</span><span class="o">.</span><span class="n">terminated</span><span class="p">:</span>
                    <span class="n">n_children</span> <span class="o">=</span> <span class="n">poisson</span><span class="p">(</span><span class="n">leaf</span><span class="o">.</span><span class="n">fitness</span><span class="p">)</span><span class="o">.</span><span class="n">rvs</span><span class="p">()</span>
                    <span class="n">leaves_unterminated</span> <span class="o">+=</span> <span class="p">(</span>
                        <span class="n">n_children</span> <span class="o">-</span> <span class="mi">1</span>
                    <span class="p">)</span>  <span class="c1"># &lt;-- this kills the parent if we drew a zero</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">n_children</span><span class="p">:</span>
                        <span class="n">leaf</span><span class="o">.</span><span class="n">terminated</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">for</span> <span class="n">child_count</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_children</span><span class="p">):</span>
                        <span class="c1"># If sequence pair mutate them separately with their own mutation rate:</span>
                        <span class="k">if</span> <span class="n">seq_bounds</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                            <span class="n">mutated_sequence1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span>
                                <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">seq_bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="p">:</span> <span class="n">seq_bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]],</span>
                                <span class="n">lambda0</span><span class="o">=</span><span class="n">lambda0</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="n">frame</span><span class="o">=</span><span class="n">frame</span><span class="p">,</span>
                            <span class="p">)</span>
                            <span class="n">mutated_sequence2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span>
                                <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="n">seq_bounds</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="p">:</span> <span class="n">seq_bounds</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]],</span>
                                <span class="n">lambda0</span><span class="o">=</span><span class="n">lambda0</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                <span class="n">frame</span><span class="o">=</span><span class="n">frame</span><span class="p">,</span>
                            <span class="p">)</span>
                            <span class="n">mutated_sequence</span> <span class="o">=</span> <span class="n">mutated_sequence1</span> <span class="o">+</span> <span class="n">mutated_sequence2</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">mutated_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutate</span><span class="p">(</span>
                                <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span><span class="p">,</span> <span class="n">lambda0</span><span class="o">=</span><span class="n">lambda0</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">frame</span><span class="o">=</span><span class="n">frame</span>
                            <span class="p">)</span>
                        <span class="n">child</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="p">()</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">dist</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">hamming_distance</span><span class="p">(</span>
                            <span class="n">mutated_sequence</span><span class="p">,</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span>
                        <span class="p">)</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">,</span> <span class="n">mutated_sequence</span><span class="p">)</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;abundance&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;terminated&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                        <span class="n">child</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;time&quot;</span><span class="p">,</span> <span class="n">t</span><span class="p">)</span>
                        <span class="n">leaf</span><span class="o">.</span><span class="n">add_child</span><span class="p">(</span><span class="n">child</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">N</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">leaves_unterminated</span> <span class="o">&lt;</span> <span class="n">N</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;tree terminated with </span><span class="si">{}</span><span class="s2"> leaves, </span><span class="si">{}</span><span class="s2"> desired&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">leaves_unterminated</span><span class="p">,</span> <span class="n">N</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># each leaf in final generation gets an observed abundance of 1, unless downsampled</span>
        <span class="k">if</span> <span class="n">T</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">T</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># Iterate the intermediate time steps:</span>
            <span class="k">for</span> <span class="n">Ti</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">T</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                <span class="c1"># Only sample those that have been &#39;sampled&#39; at intermediate sampling times:</span>
                <span class="n">final_leaves</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">leaf</span>
                    <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="n">Ti</span> <span class="ow">and</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sampled</span>
                <span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_leaves</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">n</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                        <span class="s2">&quot;tree terminated with </span><span class="si">{}</span><span class="s2"> leaves, less than what desired after downsampling </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">leaves_unterminated</span><span class="p">,</span> <span class="n">n</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">for</span> <span class="p">(</span>
                    <span class="n">leaf</span>
                <span class="p">)</span> <span class="ow">in</span> <span class="p">(</span>
                    <span class="n">final_leaves</span>
                <span class="p">):</span>  <span class="c1"># No need to down-sample, this was already done in the simulation loop</span>
                    <span class="n">leaf</span><span class="o">.</span><span class="n">abundance</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="c1"># Do the normal sampling of the last time step:</span>
        <span class="n">final_leaves</span> <span class="o">=</span> <span class="p">[</span><span class="n">leaf</span> <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">()</span> <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">time</span> <span class="o">==</span> <span class="n">t</span><span class="p">]</span>
        <span class="c1"># by default, downsample to the target simulation size</span>
        <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_leaves</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">n</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">final_leaves</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
                <span class="n">leaf</span><span class="o">.</span><span class="n">abundance</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">n</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">N</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">final_leaves</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
                <span class="n">leaf</span><span class="o">.</span><span class="n">abundance</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">N</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">T</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">final_leaves</span><span class="p">:</span>
                <span class="n">leaf</span><span class="o">.</span><span class="n">abundance</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_leaves</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">n</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;tree terminated with </span><span class="si">{}</span><span class="s2"> leaves, less than what desired after downsampling </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">leaves_unterminated</span><span class="p">,</span> <span class="n">n</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;Unknown option.&quot;</span><span class="p">)</span>

        <span class="c1"># prune away lineages that are unobserved</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">node2</span><span class="o">.</span><span class="n">abundance</span> <span class="k">for</span> <span class="n">node2</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">traverse</span><span class="p">())</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">node</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span>

        <span class="c1"># # remove unobserved unifurcations</span>
        <span class="c1"># for node in tree.iter_descendants():</span>
        <span class="c1">#     parent = node.up</span>
        <span class="c1">#     if node.abundance == 0 and len(node.children) == 1:</span>
        <span class="c1">#         node.delete(prevent_nondicotomic=False)</span>
        <span class="c1">#         node.children[0].dist = hamming_distance(node.children[0].sequence, parent.sequence)</span>

        <span class="c1"># assign unique names to each node</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">(),</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;simcell_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="c1"># return the uncollapsed tree</span>
        <span class="k">return</span> <span class="n">tree</span></div></div>


<span class="k">def</span> <span class="nf">_sequence_disambiguations</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">_accum</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Iterates through possible disambiguations of sequence, recursively.</span>

<span class="sd">    Recursion-depth-limited by number of ambiguity codes in sequence,</span>
<span class="sd">    not sequence length.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">sequence</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">base</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">bases</span><span class="p">:</span>
                <span class="n">_accum</span> <span class="o">+=</span> <span class="n">base</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">newbase</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">ambiguous_dna_values</span><span class="p">[</span><span class="n">base</span><span class="p">]:</span>
                    <span class="k">yield from</span> <span class="n">_sequence_disambiguations</span><span class="p">(</span>
                        <span class="n">sequence</span><span class="p">[</span><span class="n">index</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:],</span> <span class="n">_accum</span><span class="o">=</span><span class="p">(</span><span class="n">_accum</span> <span class="o">+</span> <span class="n">newbase</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="k">return</span>
    <span class="k">yield</span> <span class="n">_accum</span>


<span class="k">def</span> <span class="nf">_mutability_dagfuncs</span><span class="p">(</span>
    <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">splits</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Return functions for counting mutability parsimony on the history DAG.</span>

<span class="sd">    Mutability parsimony of a tree is the sum over all edges in the tree</span>
<span class="sd">    of mutability distances between parent and child node sequences.</span>

<span class="sd">    The mutability distance from an ancestral sequence to a target sequence is the sum of</span>
<span class="sd">    :math:`-log(mutability * p)` over all sites which do match, where :math:`mutability`</span>
<span class="sd">    is the mutation frequency of the k-mer surrounding the mutated base</span>
<span class="sd">    (in the ancestral sequence), and :math:`p` is the transition probability to the new</span>
<span class="sd">    base in the target sequence. Notice that this so-called distance is not symmetric.</span>

<span class="sd">    These functions may not be stable numerically, but we expect every tree to have</span>
<span class="sd">    a unique mutability parsimony score for non-degenerate mutability models, so</span>
<span class="sd">    so this shouldn&#39;t matter in practice.</span>

<span class="sd">    Args:</span>
<span class="sd">        splits: A list of indices at which sequences are concatenated, but are not</span>
<span class="sd">            adjacent in the genome.</span>
<span class="sd">        Other arguments are passed to :meth:`mutation_model.MutationModel` constructor.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A :meth:`historydag.utils.AddFuncDict` which may be passed as keyword arguments</span>
<span class="sd">        to :meth:`historydag.HistoryDag.weight_count`, :meth:`historydag.HistoryDag.trim_optimal_weight`,</span>
<span class="sd">        or :meth:`historydag.HistoryDag.optimal_weight_annotate`</span>
<span class="sd">        methods to trim or annotate a :meth:`historydag.HistoryDag` according to mutability model parsimony.</span>
<span class="sd">        Weight format is ``float``.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">mutation_model</span> <span class="o">=</span> <span class="n">MutationModel</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">dist</span> <span class="o">=</span> <span class="n">_mutability_distance</span><span class="p">(</span><span class="n">mutation_model</span><span class="p">,</span> <span class="n">splits</span><span class="o">=</span><span class="n">splits</span><span class="p">)</span>

    <span class="nd">@hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">access_field</span><span class="p">(</span><span class="s2">&quot;label&quot;</span><span class="p">)</span>
    <span class="nd">@hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">ignore_ualabel</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="nd">@hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">access_field</span><span class="p">(</span><span class="s2">&quot;sequence&quot;</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">distance</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">dist</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">hdag</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">AddFuncDict</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;start_func&quot;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;edge_weight_func&quot;</span><span class="p">:</span> <span class="n">distance</span><span class="p">,</span> <span class="s2">&quot;accum_func&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">},</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Mut. Pars.&quot;</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">_mutability_distance_precursors</span><span class="p">(</span>
    <span class="n">mutation_model</span><span class="p">:</span> <span class="n">MutationModel</span><span class="p">,</span> <span class="n">splits</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="p">):</span>
    <span class="n">chunk_idxs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">splits</span><span class="p">,</span> <span class="n">splits</span> <span class="o">+</span> <span class="p">[</span><span class="kc">None</span><span class="p">]))</span>
    <span class="c1"># Caching could be moved to the MutationModel class instead.</span>
    <span class="n">context_model</span> <span class="o">=</span> <span class="n">mutation_model</span><span class="o">.</span><span class="n">context_model</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">k</span> <span class="o">=</span> <span class="n">mutation_model</span><span class="o">.</span><span class="n">k</span>
    <span class="n">h</span> <span class="o">=</span> <span class="n">k</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="c1"># Build all sequences with (when k=5) one or two Ns on either end</span>
    <span class="n">templates</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="n">left</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="n">k</span> <span class="o">-</span> <span class="n">left</span> <span class="o">-</span> <span class="n">right</span><span class="p">),</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="n">right</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">left</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">h</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">right</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">h</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">left</span> <span class="o">!=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">right</span> <span class="o">!=</span> <span class="mi">0</span>
    <span class="p">]</span>

    <span class="n">kmers_to_compute</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">leftns</span> <span class="o">+</span> <span class="n">stub</span> <span class="o">+</span> <span class="n">rightns</span>
        <span class="k">for</span> <span class="n">leftns</span><span class="p">,</span> <span class="n">ambig_stub</span><span class="p">,</span> <span class="n">rightns</span> <span class="ow">in</span> <span class="n">templates</span>
        <span class="k">for</span> <span class="n">stub</span> <span class="ow">in</span> <span class="n">_sequence_disambiguations</span><span class="p">(</span><span class="n">ambig_stub</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="c1"># Cache all these mutabilities in context_model also</span>
    <span class="n">context_model</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span><span class="n">kmer</span><span class="p">:</span> <span class="n">mutation_model</span><span class="o">.</span><span class="n">mutability</span><span class="p">(</span><span class="n">kmer</span><span class="p">)</span> <span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">kmers_to_compute</span><span class="p">}</span>
    <span class="p">)</span>

    <span class="k">def</span> <span class="nf">add_ns</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">ns</span> <span class="o">=</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="n">h</span>
        <span class="n">chunks</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">chunk_idxs</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">ns</span> <span class="o">+</span> <span class="n">ns</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chunks</span><span class="p">)</span> <span class="o">+</span> <span class="n">ns</span>

    <span class="nd">@utils</span><span class="o">.</span><span class="n">_check_distance_arguments</span>
    <span class="k">def</span> <span class="nf">mutpairs</span><span class="p">(</span><span class="n">seq1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">seq2</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">seq1N</span> <span class="o">=</span> <span class="n">add_ns</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
        <span class="n">seq2N</span> <span class="o">=</span> <span class="n">add_ns</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
        <span class="n">mut_idxs</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">index</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="n">base1</span><span class="p">,</span> <span class="n">base2</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">seq1N</span><span class="p">,</span> <span class="n">seq2N</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">base1</span> <span class="o">!=</span> <span class="n">base2</span>
        <span class="p">]</span>
        <span class="k">return</span> <span class="n">FrozenMultiset</span><span class="p">((</span><span class="n">seq1N</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="n">h</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="n">h</span> <span class="o">+</span> <span class="mi">1</span><span class="p">],</span> <span class="n">seq2N</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mut_idxs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">sum_minus_logp</span><span class="p">(</span><span class="n">pairs</span><span class="p">:</span> <span class="n">FrozenMultiset</span><span class="p">):</span>
        <span class="c1"># I have chosen to multiply substitution rate for central base</span>
        <span class="c1"># with rate of new base. Not sure if this is a good choice.</span>
        <span class="k">if</span> <span class="n">pairs</span><span class="p">:</span>
            <span class="c1"># for floating point behavior</span>
            <span class="n">pairs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">pairs</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">p_arr</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">mult</span>
                <span class="o">*</span> <span class="p">(</span>
                    <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">context_model</span><span class="p">[</span><span class="n">mer</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                    <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">context_model</span><span class="p">[</span><span class="n">mer</span><span class="p">][</span><span class="mi">1</span><span class="p">][</span><span class="n">newbase</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="p">(</span><span class="n">mer</span><span class="p">,</span> <span class="n">newbase</span><span class="p">),</span> <span class="n">mult</span> <span class="ow">in</span> <span class="n">pairs</span>
            <span class="p">]</span>
            <span class="k">return</span> <span class="o">-</span><span class="nb">sum</span><span class="p">(</span><span class="n">p_arr</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="mf">0.0</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">mutpairs</span><span class="p">,</span> <span class="n">sum_minus_logp</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_mutability_distance</span><span class="p">(</span><span class="n">mutation_model</span><span class="p">:</span> <span class="n">MutationModel</span><span class="p">,</span> <span class="n">splits</span><span class="o">=</span><span class="p">[]):</span>
    <span class="sd">&quot;&quot;&quot;Returns a fast distance function based on passed mutation_model.</span>

<span class="sd">    First, caches computed mutabilities for k-mers with k // 2 N&#39;s on either end. This</span>
<span class="sd">    is pretty fast for k=5, but the distance function should be created once</span>
<span class="sd">    and reused.</span>

<span class="sd">    The returned distance function sums :math:`-log(mutability * p)` over all</span>
<span class="sd">    sites which do match between its two sequence arguments, where ``mutability``</span>
<span class="sd">    is the mutation frequency of the k-mer surrounding the mutated base (in the</span>
<span class="sd">    first sequence argument) and ``p`` is the transition probability to the new</span>
<span class="sd">    base.</span>

<span class="sd">    Note that, in particular, this function is not symmetric on its  arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mutpairs</span><span class="p">,</span> <span class="n">sum_minus_logp</span> <span class="o">=</span> <span class="n">_mutability_distance_precursors</span><span class="p">(</span>
        <span class="n">mutation_model</span><span class="p">,</span> <span class="n">splits</span><span class="o">=</span><span class="n">splits</span>
    <span class="p">)</span>

    <span class="k">def</span> <span class="nf">distance</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">sum_minus_logp</span><span class="p">(</span><span class="n">mutpairs</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">distance</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, William DeWitt.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>